diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/wasi/mod.rs b/src/wasi/mod.rs
--- a/src/wasi/mod.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/wasi/mod.rs	2026-08-10 23:28:48.350376110 +0200
@@ -1,5 +1,4 @@
 use super::{Send, Sync};
-use core::iter::Iterator;
 
 pub use ffi::c_void;
 
@@ -10,8 +9,15 @@
 pub type c_uint = u32;
 pub type c_short = i16;
 pub type c_ushort = u16;
-pub type c_long = i32;
-pub type c_ulong = u32;
+cfg_if! {
+    if #[cfg(target_arch = "wasm64")] {
+        pub type c_long = i64;
+        pub type c_ulong = u64;
+    } else {
+        pub type c_long = i32;
+        pub type c_ulong = u32;
+    }
+}
 pub type c_longlong = i64;
 pub type c_ulonglong = u64;
 pub type intmax_t = i64;
@@ -28,7 +34,6 @@
 pub type c_double = f64;
 pub type c_float = f32;
 pub type ino_t = u64;
-pub type sigset_t = c_uchar;
 pub type suseconds_t = c_longlong;
 pub type mode_t = u32;
 pub type dev_t = u64;
@@ -176,11 +181,6 @@
         pub st_ctim: timespec,
         __reserved: [c_longlong; 3],
     }
-
-    pub struct fd_set {
-        __nfds: usize,
-        __fds: [c_int; FD_SETSIZE as usize],
-    }
 }
 
 // Declare dirent outside of s! so that it doesn't implement Copy, Eq, Hash,
@@ -197,6 +197,9 @@
     pub d_name: [c_char; 0],
 }
 
+pub const INT_MIN: c_int = -2147483648;
+pub const INT_MAX: c_int = 2147483647;
+
 pub const EXIT_SUCCESS: c_int = 0;
 pub const EXIT_FAILURE: c_int = 1;
 pub const STDIN_FILENO: c_int = 0;
@@ -212,6 +215,8 @@
 pub const F_SETFD: c_int = 2;
 pub const F_GETFL: c_int = 3;
 pub const F_SETFL: c_int = 4;
+pub const F_DUPFD: c_int = 5;
+pub const F_DUPFD_CLOEXEC: c_int = 6;
 pub const FD_CLOEXEC: c_int = 1;
 pub const FD_SETSIZE: size_t = 1024;
 pub const O_APPEND: c_int = 0x0001;
@@ -370,28 +375,15 @@
 pub const _SC_PAGESIZE: c_int = 30;
 pub const _SC_PAGE_SIZE: ::c_int = _SC_PAGESIZE;
 pub const _SC_IOV_MAX: c_int = 60;
+pub const _SC_NPROCESSORS_ONLN: ::c_int = 84;
 pub const _SC_SYMLOOP_MAX: c_int = 173;
 
-cfg_if! {
-    if #[cfg(libc_ctest)] {
-        // skip these constants when this is active because `ctest` currently
-        // panics on parsing the constants below
-    } else {
-        // `addr_of!(EXTERN_STATIC)` is now safe; remove `unsafe` when MSRV >= 1.82
-        #[allow(unused_unsafe)]
-        pub static CLOCK_MONOTONIC: clockid_t =
-            unsafe { clockid_t(ptr_addr_of!(_CLOCK_MONOTONIC)) };
-        #[allow(unused_unsafe)]
-        pub static CLOCK_PROCESS_CPUTIME_ID: clockid_t =
-            unsafe { clockid_t(ptr_addr_of!(_CLOCK_PROCESS_CPUTIME_ID)) };
-        #[allow(unused_unsafe)]
-        pub static CLOCK_REALTIME: clockid_t =
-            unsafe { clockid_t(ptr_addr_of!(_CLOCK_REALTIME)) };
-        #[allow(unused_unsafe)]
-        pub static CLOCK_THREAD_CPUTIME_ID: clockid_t =
-            unsafe { clockid_t(ptr_addr_of!(_CLOCK_THREAD_CPUTIME_ID)) };
-    }
-}
+pub static CLOCK_MONOTONIC: clockid_t = unsafe { clockid_t(ptr_addr_of!(_CLOCK_MONOTONIC)) };
+pub static CLOCK_PROCESS_CPUTIME_ID: clockid_t =
+    unsafe { clockid_t(ptr_addr_of!(_CLOCK_PROCESS_CPUTIME_ID)) };
+pub static CLOCK_REALTIME: clockid_t = unsafe { clockid_t(ptr_addr_of!(_CLOCK_REALTIME)) };
+pub static CLOCK_THREAD_CPUTIME_ID: clockid_t =
+    unsafe { clockid_t(ptr_addr_of!(_CLOCK_THREAD_CPUTIME_ID)) };
 
 pub const ABDAY_1: ::nl_item = 0x20000;
 pub const ABDAY_2: ::nl_item = 0x20001;
@@ -458,28 +450,6 @@
 pub const YESSTR: ::nl_item = 0x50002;
 pub const NOSTR: ::nl_item = 0x50003;
 
-f! {
-    pub fn FD_ISSET(fd: ::c_int, set: *const fd_set) -> bool {
-        let set = &*set;
-        let n = set.__nfds;
-        return set.__fds[..n].iter().any(|p| *p == fd)
-    }
-
-    pub fn FD_SET(fd: ::c_int, set: *mut fd_set) -> () {
-        let set = &mut *set;
-        let n = set.__nfds;
-        if !set.__fds[..n].iter().any(|p| *p == fd) {
-            set.__nfds = n + 1;
-            set.__fds[n] = fd;
-        }
-    }
-
-    pub fn FD_ZERO(set: *mut fd_set) -> () {
-        (*set).__nfds = 0;
-        return
-    }
-}
-
 #[cfg_attr(
     feature = "rustc-dep-of-std",
     link(
@@ -775,14 +745,6 @@
     pub fn nl_langinfo(item: ::nl_item) -> *mut ::c_char;
     pub fn nl_langinfo_l(item: ::nl_item, loc: ::locale_t) -> *mut ::c_char;
 
-    pub fn select(
-        nfds: c_int,
-        readfds: *mut fd_set,
-        writefds: *mut fd_set,
-        errorfds: *mut fd_set,
-        timeout: *const timeval,
-    ) -> c_int;
-
     pub fn __wasilibc_register_preopened_fd(fd: c_int, path: *const c_char) -> c_int;
     pub fn __wasilibc_fd_renumber(fd: c_int, newfd: c_int) -> c_int;
     pub fn __wasilibc_unlinkat(fd: c_int, path: *const c_char) -> c_int;
@@ -880,9 +842,5 @@
     pub fn __errno_location() -> *mut ::c_int;
 }
 
-cfg_if! {
-    if #[cfg(target_env = "p2")] {
-        mod p2;
-        pub use self::p2::*;
-    }
-}
+mod wasix;
+pub use self::wasix::*;
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/wasi/p2.rs b/src/wasi/p2.rs
--- a/src/wasi/p2.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/wasi/p2.rs	2006-07-24 03:21:28.000000000 +0200
@@ -1,6 +1,11 @@
 pub type sa_family_t = ::c_ushort;
 pub type in_port_t = ::c_ushort;
 pub type in_addr_t = ::c_uint;
+pub type sa_family_t = u16;
+pub type sa_type_t = u16;
+pub type pthread_t = ::c_ulong;
+pub type pthread_key_t = ::c_uint;
+pub type sighandler_t = ::size_t;
 
 pub type socklen_t = ::c_uint;
 
