diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/wasi/mod.rs b/src/wasi/mod.rs
--- a/src/wasi/mod.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/wasi/mod.rs	2026-08-10 14:38:06.917180997 +0200
@@ -28,6 +28,7 @@
 pub type c_double = f64;
 pub type c_float = f32;
 pub type ino_t = u64;
+#[cfg(not(target_vendor = "wasmer"))]
 pub type sigset_t = c_uchar;
 pub type suseconds_t = c_longlong;
 pub type mode_t = u32;
@@ -212,6 +213,10 @@
 pub const F_SETFD: c_int = 2;
 pub const F_GETFL: c_int = 3;
 pub const F_SETFL: c_int = 4;
+#[cfg(target_vendor = "wasmer")]
+pub const F_DUPFD: c_int = 5;
+#[cfg(target_vendor = "wasmer")]
+pub const F_DUPFD_CLOEXEC: c_int = 6;
 pub const FD_CLOEXEC: c_int = 1;
 pub const FD_SETSIZE: size_t = 1024;
 pub const O_APPEND: c_int = 0x0001;
@@ -459,12 +464,14 @@
 pub const NOSTR: ::nl_item = 0x50003;
 
 f! {
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn FD_ISSET(fd: ::c_int, set: *const fd_set) -> bool {
         let set = &*set;
         let n = set.__nfds;
         return set.__fds[..n].iter().any(|p| *p == fd)
     }
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn FD_SET(fd: ::c_int, set: *mut fd_set) -> () {
         let set = &mut *set;
         let n = set.__nfds;
@@ -474,6 +481,7 @@
         }
     }
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn FD_ZERO(set: *mut fd_set) -> () {
         (*set).__nfds = 0;
         return
@@ -775,6 +784,7 @@
     pub fn nl_langinfo(item: ::nl_item) -> *mut ::c_char;
     pub fn nl_langinfo_l(item: ::nl_item, loc: ::locale_t) -> *mut ::c_char;
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn select(
         nfds: c_int,
         readfds: *mut fd_set,
@@ -881,7 +891,10 @@
 }
 
 cfg_if! {
-    if #[cfg(target_env = "p2")] {
+    if #[cfg(target_vendor = "wasmer")] {
+        mod wasix;
+        pub use self::wasix::*;
+    } else if #[cfg(target_env = "p2")] {
         mod p2;
         pub use self::p2::*;
     }
