diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/wasi/mod.rs b/src/wasi/mod.rs
--- a/src/wasi/mod.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/wasi/mod.rs	2026-08-10 14:35:47.815338508 +0200
@@ -32,6 +32,7 @@
 pub type c_double = f64;
 pub type c_float = f32;
 pub type ino_t = u64;
+#[cfg(not(target_vendor = "wasmer"))]
 pub type sigset_t = c_uchar;
 pub type suseconds_t = c_longlong;
 pub type mode_t = u32;
@@ -181,6 +182,7 @@
         __reserved: [c_longlong; 3],
     }
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub struct fd_set {
         __nfds: usize,
         __fds: [c_int; FD_SETSIZE as usize],
@@ -216,6 +218,10 @@
 pub const F_SETFD: c_int = 2;
 pub const F_GETFL: c_int = 3;
 pub const F_SETFL: c_int = 4;
+#[cfg(target_vendor = "wasmer")]
+pub const F_DUPFD: c_int = 5;
+#[cfg(target_vendor = "wasmer")]
+pub const F_DUPFD_CLOEXEC: c_int = 6;
 pub const FD_CLOEXEC: c_int = 1;
 pub const FD_SETSIZE: size_t = 1024;
 pub const O_APPEND: c_int = 0x0001;
@@ -463,12 +469,14 @@
 pub const NOSTR: crate::nl_item = 0x50003;
 
 f! {
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn FD_ISSET(fd: c_int, set: *const fd_set) -> bool {
         let set = &*set;
         let n = set.__nfds;
         return set.__fds[..n].iter().any(|p| *p == fd);
     }
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn FD_SET(fd: c_int, set: *mut fd_set) -> () {
         let set = &mut *set;
         let n = set.__nfds;
@@ -478,6 +486,7 @@
         }
     }
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn FD_ZERO(set: *mut fd_set) -> () {
         (*set).__nfds = 0;
         return;
@@ -764,6 +773,7 @@
     pub fn nl_langinfo(item: crate::nl_item) -> *mut c_char;
     pub fn nl_langinfo_l(item: crate::nl_item, loc: crate::locale_t) -> *mut c_char;
 
+    #[cfg(not(target_vendor = "wasmer"))]
     pub fn select(
         nfds: c_int,
         readfds: *mut fd_set,
@@ -870,7 +880,10 @@
 }
 
 cfg_if! {
-    if #[cfg(target_env = "p2")] {
+    if #[cfg(target_vendor = "wasmer")] {
+        mod wasix;
+        pub use self::wasix::*;
+    } else if #[cfg(target_env = "p2")] {
         mod p2;
         pub use self::p2::*;
     }
