--- a/src/sys/wasi.rs	2026-08-21 11:42:14.334007400 +0200
+++ b/src/sys/wasi.rs	2026-08-21 11:42:14.337007341 +0200
@@ -122,6 +122,14 @@
 
 unsafe impl<'a> Sync for MaybeUninitSlice<'a> {}
 
+impl std::fmt::Debug for RecvFlags {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("RecvFlags")
+            .field("is_truncated", &self.is_truncated())
+            .finish()
+    }
+}
+
 impl<'a> MaybeUninitSlice<'a> {
     pub(crate) fn new(buf: &'a mut [MaybeUninit<u8>]) -> MaybeUninitSlice<'a> {
         MaybeUninitSlice {
