diff --git a/Zend/zend_alloc.c b/Zend/zend_alloc.c
index f3dfe0f9..814cf9bb 100644
--- a/Zend/zend_alloc.c
+++ b/Zend/zend_alloc.c
@@ -728,6 +728,7 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
 
 static zend_always_inline void zend_mm_hugepage(void* ptr, size_t size)
 {
+#ifndef __wasi__
 #if defined(MADV_HUGEPAGE)
 	(void)madvise(ptr, size, MADV_HUGEPAGE);
 #elif defined(HAVE_MEMCNTL)
@@ -736,6 +737,7 @@ static zend_always_inline void zend_mm_hugepage(void* ptr, size_t size)
 #elif !defined(VM_FLAGS_SUPERPAGE_SIZE_2MB) && !defined(MAP_ALIGNED_SUPER)
 	zend_error_noreturn(E_ERROR, "huge_pages: thp unsupported on this platform");
 #endif
+#endif
 }
 
 static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
@@ -3279,9 +3281,15 @@ static void alloc_globals_ctor(zend_alloc_globals *alloc_globals)
 	char *tmp;
 
 #if ZEND_MM_CUSTOM
+#ifndef __wasi__
 	tmp = getenv("USE_ZEND_ALLOC");
 	if (tmp && !ZEND_ATOL(tmp)) {
+#endif
+#ifdef __wasi__
+		bool tracked = true;
+#else
 		bool tracked = (tmp = getenv("USE_TRACKED_ALLOC")) && ZEND_ATOL(tmp);
+#endif
 		zend_mm_heap *mm_heap = alloc_globals->mm_heap = malloc(sizeof(zend_mm_heap));
 		memset(mm_heap, 0, sizeof(zend_mm_heap));
 		mm_heap->use_custom_heap = ZEND_MM_CUSTOM_HEAP_STD;
@@ -3302,7 +3310,9 @@ static void alloc_globals_ctor(zend_alloc_globals *alloc_globals)
 			zend_hash_init(mm_heap->tracked_allocs, 1024, NULL, NULL, 1);
 		}
 		return;
+#ifndef __wasi__
 	}
+#endif
 #endif
 
 	tmp = getenv("USE_ZEND_ALLOC_HUGE_PAGES");
diff --git a/Zend/zend_extensions.c b/Zend/zend_extensions.c
index a4e5a38f..347680a7 100644
--- a/Zend/zend_extensions.c
+++ b/Zend/zend_extensions.c
@@ -28,7 +28,7 @@ static int last_resource_number;
 
 zend_result zend_load_extension(const char *path)
 {
-#if ZEND_EXTENSIONS_SUPPORT
+#if ZEND_EXTENSIONS_SUPPORT && !defined(__wasi__)
 	DL_HANDLE handle;
 
 	handle = DL_LOAD(path);
@@ -62,7 +62,7 @@ zend_result zend_load_extension(const char *path)
 
 zend_result zend_load_extension_handle(DL_HANDLE handle, const char *path)
 {
-#if ZEND_EXTENSIONS_SUPPORT
+#if ZEND_EXTENSIONS_SUPPORT && !defined(__wasi__)
 	zend_extension *new_extension;
 
 	const zend_extension_version_info *extension_version_info = (const zend_extension_version_info *) DL_FETCH_SYMBOL(handle, "extension_version_info");
@@ -228,7 +228,7 @@ void zend_shutdown_extensions(void)
 
 void zend_extension_dtor(zend_extension *extension)
 {
-#if ZEND_EXTENSIONS_SUPPORT && !ZEND_DEBUG
+#if ZEND_EXTENSIONS_SUPPORT && !ZEND_DEBUG && !defined(__wasi__)
 	if (extension->handle && !getenv("ZEND_DONT_UNLOAD_MODULES")) {
 		DL_UNLOAD(extension->handle);
 	}
diff --git a/Zend/zend_fibers.c b/Zend/zend_fibers.c
index d571a622..4906e4cb 100644
--- a/Zend/zend_fibers.c
+++ b/Zend/zend_fibers.c
@@ -261,7 +261,7 @@ static zend_fiber_stack *zend_fiber_stack_allocate(size_t size)
 
 	zend_mmap_set_name(pointer, alloc_size, "zend_fiber_stack");
 
-# if ZEND_FIBER_GUARD_PAGES
+# if ZEND_FIBER_GUARD_PAGES && !defined(__wasi__)
 	if (mprotect(pointer, ZEND_FIBER_GUARD_PAGES * page_size, PROT_NONE) < 0) {
 		zend_throw_exception_ex(NULL, 0, "Fiber stack protect failed: mprotect failed: %s (%d)", strerror(errno), errno);
 		munmap(pointer, alloc_size);
@@ -436,7 +436,7 @@ ZEND_API zend_result zend_fiber_init_context(zend_fiber_context *context, void *
 	makecontext(handle, (void (*)(void)) zend_fiber_trampoline, 0);
 
 	context->handle = handle;
-#else
+#elif !defined(__wasi__)
 	// Stack grows down, calculate the top of the stack. make_fcontext then shifts pointer to lower 16-byte boundary.
 	void *stack = (void *) ((uintptr_t) context->stack->pointer + context->stack->size);
 
@@ -449,6 +449,8 @@ ZEND_API zend_result zend_fiber_init_context(zend_fiber_context *context, void *
 
 	context->handle = make_fcontext(stack, context->stack->size, zend_fiber_trampoline);
 	ZEND_ASSERT(context->handle != NULL && "make_fcontext() never returns NULL");
+#else
+	return FAILURE;
 #endif
 
 	context->kind = kind;
@@ -523,16 +525,18 @@ ZEND_API void zend_fiber_switch_context(zend_fiber_transfer *transfer)
 
 	/* Copy transfer struct because it might live on the other fiber's stack that will eventually be destroyed. */
 	*transfer = *transfer_data;
-#else
+#elif !defined(__wasi__)
 	boost_context_data data = jump_fcontext(to->handle, transfer);
 
 	/* Copy transfer struct because it might live on the other fiber's stack that will eventually be destroyed. */
 	*transfer = *data.transfer;
+#else
+	return;
 #endif
 
 	to = transfer->context;
 
-#ifndef ZEND_FIBER_UCONTEXT
+#if !defined(ZEND_FIBER_UCONTEXT) && !defined(__wasi__)
 	/* Get the context that resumed us and update its handle to allow for symmetric coroutines. */
 	to->handle = data.handle;
 #endif
diff --git a/Zend/zend_long.h b/Zend/zend_long.h
index 3796f1c5..ea979e35 100644
--- a/Zend/zend_long.h
+++ b/Zend/zend_long.h
@@ -23,7 +23,7 @@
 #include <stdint.h>
 
 /* This is the heart of the whole int64 enablement in zval. */
-#if defined(__x86_64__) || defined(__LP64__) || defined(_LP64) || defined(_WIN64)
+#if defined(__x86_64__) || defined(__LP64__) || defined(_LP64) || defined(_WIN64) || defined(WASIX_64BIT_LONG_PATCH)
 # define ZEND_ENABLE_ZVAL_LONG64 1
 #endif
 
diff --git a/Zend/zend_portability.h b/Zend/zend_portability.h
index 6546ebfb..7c63a347 100644
--- a/Zend/zend_portability.h
+++ b/Zend/zend_portability.h
@@ -195,6 +195,9 @@
 # define DL_UNLOAD					FreeLibrary
 # define DL_HANDLE					HMODULE
 # define ZEND_EXTENSIONS_SUPPORT	1
+#elif defined(__wasi__)
+# define DL_HANDLE					void *
+# define ZEND_EXTENSIONS_SUPPORT	1
 #else
 # define DL_HANDLE					void *
 # define ZEND_EXTENSIONS_SUPPORT	0
diff --git a/Zend/zend_virtual_cwd.h b/Zend/zend_virtual_cwd.h
index 21735f6d..a149429f 100644
--- a/Zend/zend_virtual_cwd.h
+++ b/Zend/zend_virtual_cwd.h
@@ -59,6 +59,10 @@
 #include <direct.h>
 #endif
 
+#ifdef __wasi__
+#define chown(a,b,c) 0
+#endif
+
 #if defined(__osf__) || defined(_AIX)
 #include <errno.h>
 #endif
diff --git a/build/php.m4 b/build/php.m4
index 0a2169f6..bbbaba5a 100644
--- a/build/php.m4
+++ b/build/php.m4
@@ -734,7 +734,7 @@ AC_DEFUN([PHP_REQUIRE_CXX], [
 AS_VAR_IF([php_cxx_done],, [
   AC_PROG_CXX
   AC_PROG_CXXCPP
-  PHP_ADD_LIBRARY([stdc++])
+  PHP_ADD_LIBRARY([c++])
   php_cxx_done=yes
 ])
 ])
@@ -1366,7 +1366,7 @@ int main(void) {
   [php_cv_type_cookie_off64_t=yes],
   [php_cv_type_cookie_off64_t=no],
   [AS_CASE([$host_alias],
-    [*linux*], [php_cv_type_cookie_off64_t=yes],
+    [*linux*|*wasm*], [php_cv_type_cookie_off64_t=yes],
     [php_cv_type_cookie_off64_t=no])]
   )])
   AS_VAR_IF([php_cv_type_cookie_off64_t], [yes],
diff --git a/configure.ac b/configure.ac
index f96be617..49a3a6d4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -190,6 +190,8 @@ AS_VAR_IF([php_cv_have_re2c_cgoto], [yes],
   [AS_VAR_APPEND([RE2C_FLAGS], [" -g"])])
 ])
 
+AC_CHECK_DECLS([__wasi__])
+
 dnl Platform-specific compile settings.
 dnl ----------------------------------------------------------------------------
 
@@ -672,7 +674,7 @@ int main(void) {
   [php_cv_func_getaddrinfo=yes],
   [php_cv_func_getaddrinfo=no],
   [AS_CASE([$host_alias],
-    [*linux*|*midipix], [php_cv_func_getaddrinfo=yes],
+    [*linux*|*midipix|*wasm*], [php_cv_func_getaddrinfo=yes],
     [php_cv_func_getaddrinfo=no])])],
 [php_cv_func_getaddrinfo=no])])
 AS_VAR_IF([php_cv_func_getaddrinfo], [yes],
@@ -1297,10 +1299,12 @@ if test "$fiber_asm" = 'yes'; then
   AC_MSG_RESULT([$fiber_asm_file])
 else
   AS_VAR_IF([fiber_os], [mac], [AC_DEFINE([_XOPEN_SOURCE], [1], [ ])])
-  AC_CHECK_HEADER([ucontext.h],
-    [AC_DEFINE([ZEND_FIBER_UCONTEXT], [1],
-      [Define to 1 if Zend fiber uses ucontext instead of boost context.])],
-    [AC_MSG_FAILURE([fibers not available on this platform])])
+  AS_IF([test "$ac_cv_have_decl___wasi__" != yes], [
+    AC_CHECK_HEADER([ucontext.h],
+      [AC_DEFINE([ZEND_FIBER_UCONTEXT], [1],
+        [Define to 1 if Zend fiber uses ucontext instead of boost context.])],
+      [AC_MSG_FAILURE([fibers not available on this platform])])
+  ])
 fi
 
 ZEND_INIT
diff --git a/ext/curl/config.m4 b/ext/curl/config.m4
index bbb41fdd..58e81d82 100644
--- a/ext/curl/config.m4
+++ b/ext/curl/config.m4
@@ -11,7 +11,11 @@ if test "$PHP_CURL" != "no"; then
   PHP_EVAL_INCLINE([$CURL_CFLAGS])
 
   AC_MSG_CHECKING([for SSL support in libcurl])
-  AS_CASE([$CURL_FEATURES], [*SSL*], [CURL_SSL=yes], [CURL_SSL=no])
+  AS_IF([test "$ac_cv_have_decl___wasi__" = yes], [
+    CURL_SSL=yes
+  ], [
+    AS_CASE([$CURL_FEATURES], [*SSL*], [CURL_SSL=yes], [CURL_SSL=no])
+  ])
   AC_MSG_RESULT([$CURL_SSL])
 
   AS_IF([test "x$PHP_THREAD_SAFETY" = xyes && test "x$CURL_SSL" = xyes],
diff --git a/ext/curl/interface.c b/ext/curl/interface.c
index 5f69beac..19c161c4 100644
--- a/ext/curl/interface.c
+++ b/ext/curl/interface.c
@@ -223,7 +223,7 @@ static inline zend_result build_mime_structure_from_hash(php_curl *ch, zval *zpo
 
 /* {{{ PHP_INI_BEGIN */
 PHP_INI_BEGIN()
-	PHP_INI_ENTRY("curl.cainfo", "", PHP_INI_SYSTEM, NULL)
+	PHP_INI_ENTRY("curl.cainfo", "/etc/ssl/cacert.pem", PHP_INI_SYSTEM, NULL)
 PHP_INI_END()
 /* }}} */
 
diff --git a/ext/date/lib/timelib.h b/ext/date/lib/timelib.h
index acae8e32..7d4027ca 100644
--- a/ext/date/lib/timelib.h
+++ b/ext/date/lib/timelib.h
@@ -109,7 +109,7 @@ typedef unsigned __int64  uint64_t;
 # endif
 #endif
 
-#if (defined(__x86_64__) || defined(__LP64__) || defined(_LP64) || defined(_WIN64)) && !defined(TIMELIB_FORCE_LONG32)
+#if (defined(__x86_64__) || defined(__LP64__) || defined(_LP64) || defined(_WIN64) || defined(WASIX_64BIT_LONG_PATCH)) && !defined(TIMELIB_FORCE_LONG32)
 typedef int64_t timelib_long;
 typedef uint64_t timelib_ulong;
 # define TIMELIB_LONG_MAX INT64_MAX
diff --git a/ext/fileinfo/libmagic/fsmagic.c b/ext/fileinfo/libmagic/fsmagic.c
index 661fdb9f..14024933 100644
--- a/ext/fileinfo/libmagic/fsmagic.c
+++ b/ext/fileinfo/libmagic/fsmagic.c
@@ -167,7 +167,7 @@ file_fsmagic(struct magic_set *ms, const char *fn, zend_stat_t *sb)
 # endif
 #endif
 
-#ifdef	S_IFIFO
+#if	defined(S_IFIFO) && !defined(__wasi__)
 	case S_IFIFO:
 		if((ms->flags & MAGIC_DEVICES) != 0)
 			break;
diff --git a/ext/mysqlnd/mysqlnd_connection.c b/ext/mysqlnd/mysqlnd_connection.c
index 9065aaf5..c64102bb 100644
--- a/ext/mysqlnd/mysqlnd_connection.c
+++ b/ext/mysqlnd/mysqlnd_connection.c
@@ -526,6 +526,13 @@ MYSQLND_METHOD(mysqlnd_conn_data, get_scheme)(MYSQLND_CONN_DATA * conn, MYSQLND_
 	DBG_ENTER("mysqlnd_conn_data::get_scheme");
 #ifndef PHP_WIN32
 	if (hostname.l == sizeof("localhost") - 1 && !strncasecmp(hostname.s, "localhost", hostname.l)) {
+#ifdef __wasi__
+		// No unix socket support in WASIX, use 127.0.0.1 instead
+		if (!port) {
+			port = 3306;
+		}
+		transport.l = mnd_sprintf(&transport.s, 0, "tcp://127.0.0.1:%u", port);
+#else
 		DBG_INF_FMT("socket=%s", socket_or_pipe->s? socket_or_pipe->s:"n/a");
 		if (!socket_or_pipe->s) {
 			socket_or_pipe->s = "/tmp/mysql.sock";
@@ -533,6 +540,7 @@ MYSQLND_METHOD(mysqlnd_conn_data, get_scheme)(MYSQLND_CONN_DATA * conn, MYSQLND_
 		}
 		transport.l = mnd_sprintf(&transport.s, 0, "unix://%s", socket_or_pipe->s);
 		*unix_socket = TRUE;
+#endif
 #else
 	if (hostname.l == sizeof(".") - 1 && hostname.s[0] == '.') {
 		/* named pipe in socket */
diff --git a/ext/pdo_firebird/php_pdo_firebird_int.h b/ext/pdo_firebird/php_pdo_firebird_int.h
index a62c152f..350f1c17 100644
--- a/ext/pdo_firebird/php_pdo_firebird_int.h
+++ b/ext/pdo_firebird/php_pdo_firebird_int.h
@@ -47,7 +47,7 @@
 #define min(a,b) ((a)<(b)?(a):(b))
 #endif
 
-#if defined(_LP64) || defined(__LP64__) || defined(__arch64__) || defined(_WIN64)
+#if defined(_LP64) || defined(__LP64__) || defined(__arch64__) || defined(_WIN64) || defined(WASIX_64BIT_LONG_PATCH)
 # define PDO_FIREBIRD_HANDLE_INITIALIZER 0U
 #else
 # define PDO_FIREBIRD_HANDLE_INITIALIZER NULL
diff --git a/ext/random/csprng.c b/ext/random/csprng.c
index 73832d7d..06ebb5b9 100644
--- a/ext/random/csprng.c
+++ b/ext/random/csprng.c
@@ -48,7 +48,7 @@
 #ifdef HAVE_SYS_PARAM_H
 # include <sys/param.h>
 # if (defined(__FreeBSD__) && __FreeBSD_version > 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || \
-     (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000) || defined(__midipix__)
+     (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000) || defined(__midipix__) || defined(__wasi__)
 #  include <sys/random.h>
 # endif
 #endif
@@ -100,7 +100,7 @@ ZEND_ATTRIBUTE_NONNULL PHPAPI zend_result php_random_bytes_ex(void *bytes, size_
 #else
 	size_t read_bytes = 0;
 # if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || \
-  (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000) || defined(__midipix__)
+  (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000) || defined(__midipix__) || defined(__wasi__)
 	/* Linux getrandom(2) syscall or FreeBSD/DragonFlyBSD/NetBSD getrandom(2) function
 	 * Being a syscall, implemented in the kernel, getrandom offers higher quality output
 	 * compared to the arc4random api albeit a fallback to /dev/urandom is considered.
diff --git a/ext/standard/config.m4 b/ext/standard/config.m4
index ef6b3c5a..300846b0 100644
--- a/ext/standard/config.m4
+++ b/ext/standard/config.m4
@@ -276,7 +276,7 @@ dnl
 dnl Check if there is a support means of creating a new process and defining
 dnl which handles it receives
 dnl
-AC_CHECK_FUNCS([fork CreateProcess],
+AC_CHECK_FUNCS([fork posix_spawn CreateProcess],
   [php_can_support_proc_open=yes; break;],
   [php_can_support_proc_open=no])
 AC_MSG_CHECKING([if your OS can spawn processes with inherited handles])
@@ -384,6 +384,16 @@ AS_VAR_IF([php_cv_func_getifaddrs], [yes],
     [Define to 1 if you have the 'getifaddrs' function.])])
 
 dnl
+dnl
+dnl Whether to enable 64-bit long ints when building to WASIX;
+dnl wasm32 uses 32-bit pointers but supports 64-bit int operations.
+dnl
+if test "$ac_cv_have_decl___wasi__" = yes; then
+  if test "$WASIX_64BIT_LONG_PATCH" = yes; then
+    AC_DEFINE(WASIX_64BIT_LONG_PATCH, 1, [whether to use 64-bit longs for WASIX])
+  fi
+fi
+
 dnl Setup extension sources
 dnl
 PHP_NEW_EXTENSION([standard], m4_normalize([
diff --git a/ext/standard/filestat.c b/ext/standard/filestat.c
index 85c2517e..7f185494 100644
--- a/ext/standard/filestat.c
+++ b/ext/standard/filestat.c
@@ -236,6 +236,9 @@ static zend_result php_disk_free_space(char *path, double *space) /* {{{ */
 		return FAILURE;
 	}
 	bytesfree = (((double)buf.f_bsize) * ((double)buf.f_bavail));
+#elif defined(__wasi__)
+	*space = (double)(1024 * 1024 * 1024);
+	return SUCCESS;
 #endif
 
 	*space = bytesfree;
diff --git a/ext/standard/flock_compat.c b/ext/standard/flock_compat.c
index b7e1df7c..9fbdc1e7 100644
--- a/ext/standard/flock_compat.c
+++ b/ext/standard/flock_compat.c
@@ -26,7 +26,12 @@ PHPAPI int flock(int fd, int operation)
 #endif /* !defined(HAVE_FLOCK) */
 
 PHPAPI int php_flock(int fd, int operation)
-#ifdef HAVE_STRUCT_FLOCK /* {{{ */
+#if defined(__wasi__)
+{
+	errno = 0;
+	return 0;
+}
+#elif defined(HAVE_STRUCT_FLOCK) /* {{{ */
 {
 	struct flock flck;
 	int ret;
diff --git a/ext/standard/php_filestat.h b/ext/standard/php_filestat.h
index ef3c1cd6..c0daf967 100644
--- a/ext/standard/php_filestat.h
+++ b/ext/standard/php_filestat.h
@@ -37,6 +37,14 @@ PHP_RSHUTDOWN_FUNCTION(filestat);
 #define getuid() 1
 #endif
 
+#ifdef __wasi__
+#undef getgid
+#define getgroups(a, b) 0
+#define getgid() 1
+#define getuid() 1
+#define chown(a,b,c) 0
+#endif
+
 /* Compatibility. */
 typedef size_t php_stat_len;
 
diff --git a/ext/standard/php_fopen_wrapper.c b/ext/standard/php_fopen_wrapper.c
index ea33ba49..4a25b07b 100644
--- a/ext/standard/php_fopen_wrapper.c
+++ b/ext/standard/php_fopen_wrapper.c
@@ -317,7 +317,7 @@ static php_stream * php_stream_url_wrap_php(php_stream_wrapper *wrapper, const c
 			return NULL;
 		}
 
-#ifdef HAVE_UNISTD_H
+#if defined(HAVE_UNISTD_H) && !defined(__wasi__)
 		dtablesize = getdtablesize();
 #else
 		dtablesize = INT_MAX;
diff --git a/main/network.c b/main/network.c
index f613a73f..32eae02b 100644
--- a/main/network.c
+++ b/main/network.c
@@ -347,7 +347,10 @@ PHPAPI int php_network_connect_socket(php_socket_t sockfd,
 	socklen_t len;
 	int ret = 0;
 
+	// WASIX only supports blocking connects for now
+#ifndef __wasi__
 	SET_SOCKET_BLOCKING_MODE(sockfd, orig_flags);
+#endif
 
 	if ((n = connect(sockfd, addr, addrlen)) != 0) {
 		error = php_socket_errno();
@@ -430,10 +433,17 @@ PHPAPI int php_network_connect_socket(php_socket_t sockfd,
 	}
 
 ok:
+#ifdef __wasi__
+	if (asynchronous) {
+		// Switch to non-blocking mode after we're connected
+		SET_SOCKET_BLOCKING_MODE(sockfd, orig_flags);
+	}
+#else
 	if (!asynchronous) {
 		/* back to blocking mode */
 		RESTORE_SOCKET_BLOCKING_MODE(sockfd, orig_flags);
 	}
+#endif
 
 	if (error_code) {
 		*error_code = error;
diff --git a/main/php_open_temporary_file.c b/main/php_open_temporary_file.c
index 3a410eb7..65cf3efa 100644
--- a/main/php_open_temporary_file.c
+++ b/main/php_open_temporary_file.c
@@ -279,6 +279,9 @@ PHPAPI const char* php_get_temporary_directory(void)
 		free(tmp);
 		return PG(php_sys_temp_dir);
 	}
+#elif defined(__wasi__)
+	PG(php_sys_temp_dir) = estrdup("/tmp");
+	return PG(php_sys_temp_dir);
 #else
 	/* On Unix use the (usual) TMPDIR environment variable. */
 	{

diff --git a/ext/pcre/pcre2lib/sljit/sljitUtils.c b/ext/pcre/pcre2lib/sljit/sljitUtils.c
index 84f3aaf7..112560eb 100644
--- a/ext/pcre/pcre2lib/sljit/sljitUtils.c
+++ b/ext/pcre/pcre2lib/sljit/sljitUtils.c
@@ -327,10 +327,12 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_u8 *SLJIT_FUNC sljit_stack_resize(struct sljit_st
 		aligned_old_start = ((sljit_uw)stack->start) & ~page_align;
 
 		if (aligned_new_start > aligned_old_start) {
+#ifndef __wasi__
 			posix_madvise((void*)aligned_old_start, aligned_new_start - aligned_old_start, POSIX_MADV_DONTNEED);
 #ifdef MADV_FREE
 			madvise((void*)aligned_old_start, aligned_new_start - aligned_old_start, MADV_FREE);
 #endif /* MADV_FREE */
+#endif
 		}
 	}
 #endif /* _WIN32 */

diff --git a/Zend/zend_fibers.c b/Zend/zend_fibers.c
index d571a622..2d25dc8b 100644
--- a/Zend/zend_fibers.c
+++ b/Zend/zend_fibers.c
@@ -253,7 +253,7 @@ static zend_fiber_stack *zend_fiber_stack_allocate(size_t size)
 		return NULL;
 	}
 
-#if defined(MADV_NOHUGEPAGE)
+#if defined(MADV_NOHUGEPAGE) && !defined(__wasi__)
 	/* Multiple reasons to fail, ignore all errors only needed
 	 * for linux < 6.8 */
 	(void) madvise(pointer, alloc_size, MADV_NOHUGEPAGE);
