diff --git a/ext/opcache/ZendAccelerator.c b/ext/opcache/ZendAccelerator.c
index 904a9c2a..5a6f62bb 100644
--- a/ext/opcache/ZendAccelerator.c
+++ b/ext/opcache/ZendAccelerator.c
@@ -152,6 +152,12 @@ static void preload_restart(void);
 # define INCREMENT(v) InterlockedIncrement64(&ZCSG(v))
 # define DECREMENT(v) InterlockedDecrement64(&ZCSG(v))
 # define LOCKVAL(v)   (ZCSG(v))
+#elif defined(__wasi__)
+static uint32_t accel_wasi_mem_usage = 0;
+static uint32_t accel_wasi_restart_in = 0;
+# define INCREMENT(v) __atomic_add_fetch(&accel_wasi_##v, 1, __ATOMIC_SEQ_CST)
+# define DECREMENT(v) __atomic_sub_fetch(&accel_wasi_##v, 1, __ATOMIC_SEQ_CST)
+# define LOCKVAL(v)   __atomic_load_n(&accel_wasi_##v, __ATOMIC_SEQ_CST)
 #endif
 
 #define ZCG_KEY_LEN (MAXPATHLEN * 8)
@@ -275,9 +281,9 @@ static ZEND_INI_MH(accel_include_path_on_modify)
 
 static inline void accel_restart_enter(void)
 {
-#ifdef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
 	INCREMENT(restart_in);
-#else
+#elif !defined(__wasi__)
 	struct flock restart_in_progress;
 
 	restart_in_progress.l_type = F_WRLCK;
@@ -294,10 +300,10 @@ static inline void accel_restart_enter(void)
 
 static inline void accel_restart_leave(void)
 {
-#ifdef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
 	ZCSG(restart_in_progress) = false;
 	DECREMENT(restart_in);
-#else
+#elif !defined(__wasi__)
 	struct flock restart_finished;
 
 	restart_finished.l_type = F_UNLCK;
@@ -315,7 +321,9 @@ static inline void accel_restart_leave(void)
 static inline int accel_restart_is_active(void)
 {
 	if (ZCSG(restart_in_progress)) {
-#ifndef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
+		return LOCKVAL(restart_in) != 0;
+#elif !defined(__wasi__)
 		struct flock restart_check;
 
 		restart_check.l_type = F_WRLCK;
@@ -333,8 +341,6 @@ static inline int accel_restart_is_active(void)
 		} else {
 			return 1;
 		}
-#else
-		return LOCKVAL(restart_in) != 0;
 #endif
 	}
 	return 0;
@@ -343,11 +349,11 @@ static inline int accel_restart_is_active(void)
 /* Creates a read lock for SHM access */
 static inline zend_result accel_activate_add(void)
 {
-#ifdef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
 	SHM_UNPROTECT();
 	INCREMENT(mem_usage);
 	SHM_PROTECT();
-#else
+#elif !defined(__wasi__)
 	struct flock mem_usage_lock;
 
 	mem_usage_lock.l_type = F_RDLCK;
@@ -366,14 +372,14 @@ static inline zend_result accel_activate_add(void)
 /* Releases a lock for SHM access */
 static inline void accel_deactivate_sub(void)
 {
-#ifdef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
 	if (ZCG(counted)) {
 		SHM_UNPROTECT();
 		DECREMENT(mem_usage);
 		ZCG(counted) = false;
 		SHM_PROTECT();
 	}
-#else
+#elif !defined(__wasi__)
 	struct flock mem_usage_unlock;
 
 	mem_usage_unlock.l_type = F_UNLCK;
@@ -389,9 +395,9 @@ static inline void accel_deactivate_sub(void)
 
 static inline void accel_unlock_all(void)
 {
-#ifdef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
 	accel_deactivate_sub();
-#else
+#elif !defined(__wasi__)
 	if (lock_file == -1) {
 		return;
 	}
@@ -834,7 +840,7 @@ static void accel_use_shm_interned_strings(void)
 	HANDLE_UNBLOCK_INTERRUPTIONS();
 }
 
-#ifndef ZEND_WIN32
+#if !defined(ZEND_WIN32) && !defined(__wasi__)
 static inline void kill_all_lockers(struct flock *mem_usage_check)
 {
 	int tries;
@@ -901,7 +907,7 @@ static inline void kill_all_lockers(struct flock *mem_usage_check)
 
 static inline bool accel_is_inactive(void)
 {
-#ifdef ZEND_WIN32
+#if defined(ZEND_WIN32) || defined(__wasi__)
 	/* on Windows, we don't need kill_all_lockers() because SAPIs
 	   that work on Windows don't manage child processes (and we
 	   can't do anything about hanging threads anyway); therefore
@@ -912,7 +918,7 @@ static inline bool accel_is_inactive(void)
 	if (LOCKVAL(mem_usage) == 0) {
 		return true;
 	}
-#else
+#elif !defined(__wasi__)
 	struct flock mem_usage_check;
 
 	mem_usage_check.l_type = F_WRLCK;
@@ -2938,6 +2944,10 @@ static zend_result zend_accel_init_shm(void)
 	ZCSG(start_time) = zend_accel_get_time();
 	ZCSG(last_restart_time) = 0;
 	ZCSG(restart_in_progress) = false;
+#ifdef __wasi__
+	__atomic_store_n(&accel_wasi_mem_usage, 0, __ATOMIC_SEQ_CST);
+	__atomic_store_n(&accel_wasi_restart_in, 0, __ATOMIC_SEQ_CST);
+#endif
 
 	for (i = 0; i < -HT_MIN_MASK; i++) {
 		ZCSG(uninitialized_bucket)[i] = HT_INVALID_IDX;
@@ -4989,8 +4999,10 @@ static zend_result accel_finish_startup_preload_subprocess(pid_t *pid)
 
 	if (!ZCG(accel_directives).preload_user
 	 || !*ZCG(accel_directives).preload_user) {
-
 		bool sapi_requires_preload_user = !(strcmp(sapi_module.name, "cli") == 0
+#ifdef __wasi__
+		  || strcmp(sapi_module.name, "cli-server") == 0
+#endif
 		  || strcmp(sapi_module.name, "phpdbg") == 0);
 
 		if (!sapi_requires_preload_user) {
@@ -5015,6 +5027,7 @@ static zend_result accel_finish_startup_preload_subprocess(pid_t *pid)
 		return SUCCESS;
 	}
 
+#ifndef __wasi__
 	*pid = fork();
 	if (*pid == -1) {
 		zend_shared_alloc_unlock();
@@ -5039,6 +5052,14 @@ static zend_result accel_finish_startup_preload_subprocess(pid_t *pid)
 	}
 
 	return SUCCESS;
+#else
+	// Note: in WASIX, there is only one uid; it has "root" access to everything within the
+	// sandbox, but not the base OS, so it's OK to run everything with that user. That is to
+	// say, we shouldn't need to fork a new process here.
+	// It is also a fact that forking doesn't work under WASIX+EH anyway.
+	zend_accel_error(ACCEL_LOG_FATAL, "Preloading cannot fork to spawn a new process in WASIX");
+	exit(1);
+#endif
 }
 #endif /* ZEND_WIN32 */
 
diff --git a/ext/opcache/config.m4 b/ext/opcache/config.m4
index 70138726..a53316a8 100644
--- a/ext/opcache/config.m4
+++ b/ext/opcache/config.m4
@@ -229,7 +229,7 @@ int main(void) {
 [php_cv_shm_mmap_anon=yes],
 [php_cv_shm_mmap_anon=no],
 [AS_CASE([$host_alias],
-  [*linux*|*midipix], [php_cv_shm_mmap_anon=yes],
+  [*linux*|*midipix|*wasm*], [php_cv_shm_mmap_anon=yes],
   [php_cv_shm_mmap_anon=no])])])
 AS_VAR_IF([php_cv_shm_mmap_anon], [yes],
   [AC_DEFINE([HAVE_SHM_MMAP_ANON], [1],
diff --git a/ext/opcache/zend_accelerator_module.c b/ext/opcache/zend_accelerator_module.c
index 6f668af9..98105cec 100644
--- a/ext/opcache/zend_accelerator_module.c
+++ b/ext/opcache/zend_accelerator_module.c
@@ -312,7 +312,7 @@ ZEND_INI_BEGIN()
 	STD_PHP_INI_ENTRY("opcache.optimization_level"    , DEFAULT_OPTIMIZATION_LEVEL , PHP_INI_SYSTEM, OnUpdateLong, accel_directives.optimization_level,   zend_accel_globals, accel_globals)
 	STD_PHP_INI_ENTRY("opcache.opt_debug_level"       , "0"      , PHP_INI_SYSTEM, OnUpdateLong,             accel_directives.opt_debug_level,            zend_accel_globals, accel_globals)
 	STD_PHP_INI_BOOLEAN("opcache.enable_file_override"	, "0"   , PHP_INI_SYSTEM, OnUpdateBool,              accel_directives.file_override_enabled,     zend_accel_globals, accel_globals)
-	STD_PHP_INI_BOOLEAN("opcache.enable_cli"             , "0"   , PHP_INI_SYSTEM, OnUpdateBool,              accel_directives.enable_cli,                zend_accel_globals, accel_globals)
+	STD_PHP_INI_BOOLEAN("opcache.enable_cli"             , "1"   , PHP_INI_SYSTEM, OnUpdateBool,              accel_directives.enable_cli,                zend_accel_globals, accel_globals)
 	STD_PHP_INI_ENTRY("opcache.error_log"                , ""    , PHP_INI_SYSTEM, OnUpdateString,	         accel_directives.error_log,                 zend_accel_globals, accel_globals)
 	STD_PHP_INI_ENTRY("opcache.restrict_api"             , ""    , PHP_INI_SYSTEM, OnUpdateString,	         accel_directives.restrict_api,              zend_accel_globals, accel_globals)
 
diff --git a/ext/opcache/zend_file_cache.c b/ext/opcache/zend_file_cache.c
index d430f483..0f9e54ef 100644
--- a/ext/opcache/zend_file_cache.c
+++ b/ext/opcache/zend_file_cache.c
@@ -97,9 +97,15 @@ static int zend_file_cache_flock(int fd, int op)
 #elif defined(HAVE_FLOCK)
 # define zend_file_cache_flock flock
 #else
-# define LOCK_SH 0
-# define LOCK_EX 1
-# define LOCK_UN 2
+# ifndef LOCK_SH
+#  define LOCK_SH 0
+# endif
+# ifndef LOCK_EX
+#  define LOCK_EX 1
+# endif
+# ifndef LOCK_UN
+#  define LOCK_UN 2
+# endif
 static int zend_file_cache_flock(int fd, int type)
 {
 	return 0;
diff --git a/ext/opcache/zend_shared_alloc.c b/ext/opcache/zend_shared_alloc.c
index 80ef36b8..c854e80e 100644
--- a/ext/opcache/zend_shared_alloc.c
+++ b/ext/opcache/zend_shared_alloc.c
@@ -29,6 +29,7 @@
 #include <errno.h>
 #include "ZendAccelerator.h"
 #include "zend_shared_alloc.h"
+#include "zend_atomic.h"
 #ifdef HAVE_UNISTD_H
 # include <unistd.h>
 #endif
@@ -63,6 +64,14 @@ int lock_file = -1;
 static char lockfile_name[MAXPATHLEN];
 #endif
 
+#ifdef __wasi__
+# ifdef ZTS
+static MUTEX_T zend_shared_alloc_wasi_mutex;
+# else
+static zend_atomic_bool zend_shared_alloc_wasi_lock;
+# endif
+#endif
+
 static const zend_shared_memory_handler_entry handler_table[] = {
 #ifdef USE_MMAP
 	{ "mmap", &zend_alloc_mmap_handlers },
@@ -193,7 +202,15 @@ int zend_shared_alloc_startup(size_t requested_size, size_t reserved_size)
 	ZSMMG(shared_free) = requested_size - reserved_size; /* goes to tmp_shared_globals.shared_free */
 
 #ifndef ZEND_WIN32
+#ifdef __wasi__
+#ifdef ZTS
+	zend_shared_alloc_wasi_mutex = tsrm_mutex_alloc();
+#else
+	ZEND_ATOMIC_BOOL_INIT(&zend_shared_alloc_wasi_lock, false);
+#endif
+#else
 	zend_shared_alloc_create_lock(ZCG(accel_directives).lockfile_path);
+#endif
 #else
 	zend_shared_alloc_create_lock();
 #endif
@@ -337,10 +354,17 @@ void zend_shared_alloc_shutdown(void)
 	ZSMMG(shared_segments) = NULL;
 	g_shared_alloc_handler = NULL;
 #ifndef ZEND_WIN32
-	close(lock_file);
+	if (lock_file >= 0) {
+		close(lock_file);
+		lock_file = -1;
+	}
 
 # ifdef ZTS
+#  ifdef __wasi__
+	tsrm_mutex_free(zend_shared_alloc_wasi_mutex);
+#  else
 	tsrm_mutex_free(zts_lock);
+#  endif
 # endif
 #endif
 }
@@ -487,7 +511,16 @@ void zend_shared_alloc_lock(void)
 {
 	ZEND_ASSERT(!ZCG(locked));
 
-#ifndef ZEND_WIN32
+#ifdef ZEND_WIN32
+	zend_shared_alloc_lock_win32();
+#elif defined(__wasi__)
+#ifdef ZTS
+	tsrm_mutex_lock(zend_shared_alloc_wasi_mutex);
+#else
+	while (zend_atomic_bool_exchange_ex(&zend_shared_alloc_wasi_lock, true)) {
+	}
+#endif
+#else
 	struct flock mem_write_lock;
 
 	mem_write_lock.l_type = F_WRLCK;
@@ -515,8 +548,6 @@ void zend_shared_alloc_lock(void)
 		}
 		break;
 	}
-#else
-	zend_shared_alloc_lock_win32();
 #endif
 
 	ZCG(locked) = 1;
@@ -526,7 +557,7 @@ void zend_shared_alloc_unlock(void)
 {
 	ZEND_ASSERT(ZCG(locked));
 
-#ifndef ZEND_WIN32
+#if !defined(ZEND_WIN32) && !defined(__wasi__)
 	struct flock mem_write_unlock;
 
 	mem_write_unlock.l_type = F_UNLCK;
@@ -537,15 +568,21 @@ void zend_shared_alloc_unlock(void)
 
 	ZCG(locked) = 0;
 
-#ifndef ZEND_WIN32
+#ifdef ZEND_WIN32
+	zend_shared_alloc_unlock_win32();
+#elif defined(__wasi__)
+#ifdef ZTS
+	tsrm_mutex_unlock(zend_shared_alloc_wasi_mutex);
+#else
+	zend_atomic_bool_store_ex(&zend_shared_alloc_wasi_lock, false);
+#endif
+#else
 	if (fcntl(lock_file, F_SETLK, &mem_write_unlock) == -1) {
 		zend_accel_error_noreturn(ACCEL_LOG_ERROR, "Cannot remove lock - %s (%d)", strerror(errno), errno);
 	}
 #ifdef ZTS
 	tsrm_mutex_unlock(zts_lock);
 #endif
-#else
-	zend_shared_alloc_unlock_win32();
 #endif
 }
 

diff --git a/ext/opcache/ZendAccelerator.c b/ext/opcache/ZendAccelerator.c
index 904a9c2a..5a6f62bb 100644
--- a/ext/opcache/ZendAccelerator.c
+++ b/ext/opcache/ZendAccelerator.c
@@ -3219,13 +3219,13 @@ static void accel_move_code_to_huge_pages(void)
 			MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED,
 			-1, 0);
 		/* this should never happen? */
 		ZEND_ASSERT(ret != MAP_FAILED);
-#  ifdef MADV_HUGEPAGE
+#  if defined(MADV_HUGEPAGE) && !defined(__wasi__)
 		if (-1 == madvise(start, size, MADV_HUGEPAGE)) {
 			memcpy(start, mem, real_size);
 			mprotect(start, size, PROT_READ | PROT_EXEC);
 			munmap(mem, size);
 			zend_error(E_WARNING,
 				ACCELERATOR_PRODUCT_NAME " huge_code_pages: madvise(HUGEPAGE) failed: %s (%d)",
 				strerror(errno), errno);
 			return FAILURE;
 		}
