diff --git a/build/php.m4 b/build/php.m4
--- a/build/php.m4
+++ b/build/php.m4
@@ -1923,10 +1923,14 @@ AC_DEFUN([PHP_SETUP_PGSQL], [dnl
 found_pgsql=no
 dnl Set PostgreSQL installation directory if given from the configure argument.
 AS_CASE([$4], [yes], [pgsql_dir=""], [pgsql_dir=$4])
-AS_VAR_IF([pgsql_dir],,
-  [PKG_CHECK_MODULES([PGSQL], [libpq >= 10.0],
-    [found_pgsql=yes],
-    [found_pgsql=no])])
+AS_IF([test -n "$PGSQL_CFLAGS" && test -n "$PGSQL_LIBS"], [
+  found_pgsql=yes
+], [
+  AS_VAR_IF([pgsql_dir],,
+    [PKG_CHECK_MODULES([PGSQL], [libpq >= 10.0],
+      [found_pgsql=yes],
+      [found_pgsql=no])])
+])
 
 AS_VAR_IF([found_pgsql], [no], [dnl
   AC_MSG_CHECKING([for pg_config])
