diff --git a/ext/pgsql/pgsql.c b/ext/pgsql/pgsql.c
--- a/ext/pgsql/pgsql.c
+++ b/ext/pgsql/pgsql.c
@@ -410,7 +410,7 @@ static bool php_pgsql_strtok_identifier_parse(const char *str, size_t str_len)
 }
 
 #ifndef HAVE_PG_CHANGE_PASSWORD
-static PGresult *PQchangePassword(PGconn *conn, const char *user, const char *passwd)
+static PGresult *php_PQchangePassword(PGconn *conn, const char *user, const char *passwd)
 {
 	/**
 	 * It is more appropriate to let the configured password encryption algorithm
@@ -451,7 +451,7 @@ static PGresult *PQchangePassword(PGconn *conn, const char *user, const char *pa
 #endif
 
 #if !defined(HAVE_PG_SOCKET_POLL)
-static int PQsocketPoll(int socket, int read, int write, time_t timeout)
+static int php_PQsocketPoll(int socket, int read, int write, time_t timeout)
 {
 	if (!read && !write)
 		return 0;
@@ -6335,7 +6335,7 @@ PHP_FUNCTION(pg_change_password)
 		RETURN_THROWS();
 	}
 
-	pg_result = PQchangePassword(link->conn, ZSTR_VAL(user), ZSTR_VAL(passwd));
+	pg_result = php_PQchangePassword(link->conn, ZSTR_VAL(user), ZSTR_VAL(passwd));
 	RETVAL_BOOL(PQresultStatus(pg_result) == PGRES_COMMAND_OK);
 	PQclear(pg_result);
 }
@@ -6404,7 +6404,7 @@ PHP_FUNCTION(pg_socket_poll)
 		RETURN_FALSE;
 	}
 
-	RETURN_LONG((zend_long)PQsocketPoll(socket, (int)read, (int)write, (int)ts));
+	RETURN_LONG((zend_long)php_PQsocketPoll(socket, (int)read, (int)write, (int)ts));
 }
 /* }}} */
 #endif
