diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/Cargo.toml b/Cargo.toml
--- a/Cargo.toml	1970-01-01 01:00:01.000000000 +0100
+++ b/Cargo.toml	2026-08-10 23:28:13.832412539 +0200
@@ -71,7 +71,7 @@
 [features]
 all = []
 
-[target."cfg(unix)".dependencies.libc]
+[target.'cfg(any(unix, target_os = "wasi"))'.dependencies.libc]
 version = "0.2.141"
 
 [target."cfg(windows)".dependencies.windows-sys]
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/lib.rs b/src/lib.rs
--- a/src/lib.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/lib.rs	2006-07-24 03:21:28.000000000 +0200
@@ -101,7 +101,7 @@
     ($from: ty, $for: ty) => {
         impl From<$from> for $for {
             fn from(socket: $from) -> $for {
-                #[cfg(unix)]
+                #[cfg(any(unix, target_os = "wasi"))]
                 unsafe {
                     <$for>::from_raw_fd(socket.into_raw_fd())
                 }
@@ -170,9 +170,10 @@
 
 #[cfg_attr(unix, path = "sys/unix.rs")]
 #[cfg_attr(windows, path = "sys/windows.rs")]
+#[cfg_attr(target_os = "wasi", path = "sys/wasi.rs")]
 mod sys;
 
-#[cfg(not(any(windows, unix)))]
+#[cfg(not(any(windows, unix, target_os = "wasi")))]
 compile_error!("Socket2 doesn't support the compile target");
 
 use sys::c_int;
@@ -249,12 +250,12 @@
     /// Type corresponding to `SOCK_STREAM`.
     ///
     /// Used for protocols such as TCP.
-    pub const STREAM: Type = Type(sys::SOCK_STREAM);
+    pub const STREAM: Type = Type(sys::SOCK_STREAM as i32);
 
     /// Type corresponding to `SOCK_DGRAM`.
     ///
     /// Used for protocols such as UDP.
-    pub const DGRAM: Type = Type(sys::SOCK_DGRAM);
+    pub const DGRAM: Type = Type(sys::SOCK_DGRAM as i32);
 
     /// Type corresponding to `SOCK_DCCP`.
     ///
@@ -491,6 +492,7 @@
         target_os = "tvos",
         target_os = "watchos",
         target_os = "windows",
+        target_os = "wasi",
     ))]
     #[cfg_attr(
         docsrs,
@@ -507,6 +509,7 @@
             target_os = "tvos",
             target_os = "watchos",
             target_os = "windows",
+            target_os = "wasi",
         )))
     )]
     pub const fn with_interval(self, interval: Duration) -> Self {
@@ -531,6 +534,7 @@
             target_os = "ios",
             target_os = "linux",
             target_os = "macos",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "tvos",
             target_os = "watchos",
@@ -549,6 +553,7 @@
                 target_os = "ios",
                 target_os = "linux",
                 target_os = "macos",
+                target_os = "wasi",
                 target_os = "netbsd",
                 target_os = "tvos",
                 target_os = "watchos",
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/sockaddr.rs b/src/sockaddr.rs
--- a/src/sockaddr.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/sockaddr.rs	2006-07-24 03:21:28.000000000 +0200
@@ -1,6 +1,7 @@
 use std::hash::Hash;
 use std::mem::{self, size_of, MaybeUninit};
 use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};
+#[cfg(not(target_os = "wasi"))]
 use std::path::Path;
 use std::{fmt, io, ptr};
 
@@ -146,6 +147,7 @@
     /// Constructs a `SockAddr` with the family `AF_UNIX` and the provided path.
     ///
     /// Returns an error if the path is longer than `SUN_LEN`.
+    #[cfg(not(target_os = "wasi"))]
     pub fn unix<P>(path: P) -> io::Result<SockAddr>
     where
         P: AsRef<Path>,
@@ -206,7 +208,7 @@
     }
 
     /// Returns a raw pointer to the address storage.
-    #[cfg(all(unix, not(target_os = "redox")))]
+    #[cfg(all(any(unix, target_os = "wasi"), not(target_os = "redox")))]
     pub(crate) const fn as_storage_ptr(&self) -> *const sockaddr_storage {
         &self.storage
     }
@@ -231,7 +233,7 @@
                 ip,
                 port,
                 addr.sin6_flowinfo,
-                #[cfg(unix)]
+                #[cfg(any(unix, all(target_os = "wasi", target_vendor = "wasmer")))]
                 addr.sin6_scope_id,
                 #[cfg(windows)]
                 unsafe {
@@ -305,7 +307,7 @@
             storage.sin6_port = addr.port().to_be();
             storage.sin6_addr = crate::sys::to_in6_addr(addr.ip());
             storage.sin6_flowinfo = addr.flowinfo();
-            #[cfg(unix)]
+            #[cfg(any(unix, all(target_os = "wasi", target_vendor = "wasmer")))]
             {
                 storage.sin6_scope_id = addr.scope_id();
             }
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/socket.rs b/src/socket.rs
--- a/src/socket.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/socket.rs	2006-07-24 03:21:28.000000000 +0200
@@ -18,6 +18,8 @@
 use std::os::unix::io::{FromRawFd, IntoRawFd};
 #[cfg(windows)]
 use std::os::windows::io::{FromRawSocket, IntoRawSocket};
+#[cfg(target_os = "wasi")]
+use std::os::wasi::io::{FromRawFd, IntoRawFd};
 use std::time::Duration;
 
 use crate::sys::{self, c_int, getsockopt, setsockopt, Bool};
@@ -100,7 +102,7 @@
                 // Violating this assumption (fd never negative) causes UB,
                 // something we don't want. So check for that we have this
                 // `assert!`.
-                #[cfg(unix)]
+                #[cfg(any(unix, target_os = "wasi"))]
                 assert!(raw >= 0, "tried to create a `Socket` with an invalid fd");
                 sys::socket_from_raw(raw)
             },
@@ -231,7 +233,7 @@
         match res {
             Ok(()) => return Ok(()),
             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {}
-            #[cfg(unix)]
+            #[cfg(any(unix, target_os = "wasi"))]
             Err(ref e) if e.raw_os_error() == Some(libc::EINPROGRESS) => {}
             Err(e) => return Err(e),
         }
@@ -269,6 +271,7 @@
             target_os = "fuchsia",
             target_os = "illumos",
             target_os = "linux",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "openbsd",
         ))]
@@ -282,6 +285,7 @@
             target_os = "fuchsia",
             target_os = "illumos",
             target_os = "linux",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "openbsd",
         )))]
@@ -623,6 +627,7 @@
     /// but suppresses the `WSAEMSGSIZE` error on Windows.
     ///
     /// [`peek_from`]: Socket::peek_from
+    #[cfg(not(target_os = "wasi"))]
     pub fn peek_sender(&self) -> io::Result<SockAddr> {
         sys::peek_sender(self.as_raw())
     }
@@ -1089,7 +1094,7 @@
     #[cfg_attr(docsrs, doc(cfg(all(feature = "all", not(target_os = "redox")))))]
     pub fn header_included(&self) -> io::Result<bool> {
         unsafe {
-            getsockopt::<c_int>(self.as_raw(), sys::IPPROTO_IP, sys::IP_HDRINCL)
+            getsockopt::<c_int>(self.as_raw(), sys::IPPROTO_IP, libc::IP_HDRINCL)
                 .map(|included| included != 0)
         }
     }
@@ -1116,7 +1121,7 @@
             setsockopt(
                 self.as_raw(),
                 sys::IPPROTO_IP,
-                sys::IP_HDRINCL,
+                libc::IP_HDRINCL,
                 included as c_int,
             )
         }
@@ -1732,6 +1737,7 @@
         target_os = "redox",
         target_os = "solaris",
         target_os = "haiku",
+        target_os = "wasi",
     )))]
     pub fn recv_tclass_v6(&self) -> io::Result<bool> {
         unsafe {
@@ -1754,6 +1760,7 @@
         target_os = "redox",
         target_os = "solaris",
         target_os = "haiku",
+        target_os = "wasi",
     )))]
     pub fn set_recv_tclass_v6(&self, recv_tclass: bool) -> io::Result<()> {
         unsafe {
@@ -1808,6 +1815,7 @@
             target_os = "ios",
             target_os = "linux",
             target_os = "macos",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "tvos",
             target_os = "watchos",
@@ -1826,6 +1834,7 @@
                 target_os = "ios",
                 target_os = "linux",
                 target_os = "macos",
+                target_os = "wasi",
                 target_os = "netbsd",
                 target_os = "tvos",
                 target_os = "watchos",
@@ -1855,6 +1864,7 @@
             target_os = "ios",
             target_os = "linux",
             target_os = "macos",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "tvos",
             target_os = "watchos",
@@ -1873,6 +1883,7 @@
                 target_os = "ios",
                 target_os = "linux",
                 target_os = "macos",
+                target_os = "wasi",
                 target_os = "netbsd",
                 target_os = "tvos",
                 target_os = "watchos",
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/sockref.rs b/src/sockref.rs
--- a/src/sockref.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/sockref.rs	2006-07-24 03:21:28.000000000 +0200
@@ -6,6 +6,8 @@
 use std::os::unix::io::{AsFd, AsRawFd, FromRawFd};
 #[cfg(windows)]
 use std::os::windows::io::{AsRawSocket, AsSocket, FromRawSocket};
+#[cfg(target_os = "wasi")]
+use std::os::wasi::io::{AsFd, AsRawFd, FromRawFd};
 
 use crate::Socket;
 
@@ -77,8 +79,8 @@
 }
 
 /// On Windows, a corresponding `From<&impl AsSocket>` implementation exists.
-#[cfg(unix)]
-#[cfg_attr(docsrs, doc(cfg(unix)))]
+#[cfg(any(unix, target_os = "wasi"))]
+#[cfg_attr(docsrs, doc(cfg(any(unix, target_os = "wasi"))))]
 impl<'s, S> From<&'s S> for SockRef<'s>
 where
     S: AsFd,
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/sys/unix.rs b/src/sys/unix.rs
--- a/src/sys/unix.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/sys/unix.rs	2006-07-24 03:21:28.000000000 +0200
@@ -293,6 +293,10 @@
 #[cfg(all(feature = "all", any(target_os = "freebsd", target_os = "linux")))]
 const TCP_CA_NAME_MAX: usize = 16;
 
+// TCP_CA_NAME_MAX isn't defined in user space include files(not in libc)
+#[cfg(all(feature = "all", any(target_os = "freebsd", target_os = "linux")))]
+const TCP_CA_NAME_MAX: usize = 16;
+
 #[cfg(any(
     all(
         target_os = "linux",
@@ -1162,7 +1166,7 @@
     syscall!(fcntl(fd, cmd))
 }
 
-/// Add `flag` to the current set flags of `F_GETFD`.
+//// Add `flag` to the current set flags of `F_GETFD`.
 fn fcntl_add(fd: Socket, get_cmd: c_int, set_cmd: c_int, flag: c_int) -> io::Result<()> {
     let previous = fcntl_get(fd, get_cmd)?;
     let new = previous | flag;
@@ -2886,4 +2890,4 @@
     let want = [32, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7];
     assert_eq!(raw.s6_addr, want);
     assert_eq!(from_in6_addr(raw), ip);
-}
+}
\ No newline at end of file
