diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/Cargo.toml b/Cargo.toml
--- a/Cargo.toml	1970-01-01 01:00:01.000000000 +0100
+++ b/Cargo.toml	2026-07-31 10:37:09.973446688 +0200
@@ -50,16 +50,6 @@
 license = "MIT OR Apache-2.0"
 repository = "https://github.com/rust-lang/socket2"
 
-[package.metadata.cargo_check_external_types]
-allowed_external_types = [
-    "libc::socklen_t",
-    "libc::*::socklen_t",
-    "libc::sa_family_t",
-    "libc::*::sa_family_t",
-    "windows_sys::Win32::Networking::WinSock::socklen_t",
-    "windows_sys::Win32::Networking::WinSock::ADDRESS_FAMILY",
-]
-
 [package.metadata.docs.rs]
 all-features = true
 default-target = "x86_64-unknown-linux-gnu"
@@ -83,6 +73,16 @@
 [package.metadata.playground]
 features = ["all"]
 
+[package.metadata.cargo_check_external_types]
+allowed_external_types = [
+    "libc::socklen_t",
+    "libc::*::socklen_t",
+    "libc::sa_family_t",
+    "libc::*::sa_family_t",
+    "windows_sys::Win32::Networking::WinSock::socklen_t",
+    "windows_sys::Win32::Networking::WinSock::ADDRESS_FAMILY",
+]
+
 [features]
 all = []
 
@@ -90,8 +90,8 @@
 name = "socket2"
 path = "src/lib.rs"
 
-[target."cfg(unix)".dependencies.libc]
-version = "0.2.172"
+[target.'cfg(any(unix, target_os = "wasi"))'.dependencies.libc]
+version = "0.2.169"
 
 [target."cfg(windows)".dependencies.windows-sys]
 version = "0.59"
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/src/lib.rs b/src/lib.rs
--- a/src/lib.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/lib.rs	2006-07-24 03:21:28.000000000 +0200
@@ -109,7 +109,7 @@
     ($from: ty, $for: ty) => {
         impl From<$from> for $for {
             fn from(socket: $from) -> $for {
-                #[cfg(unix)]
+                #[cfg(any(unix, target_os = "wasi"))]
                 unsafe {
                     <$for>::from_raw_fd(socket.into_raw_fd())
                 }
@@ -178,9 +178,10 @@
 
 #[cfg_attr(unix, path = "sys/unix.rs")]
 #[cfg_attr(windows, path = "sys/windows.rs")]
+#[cfg_attr(target_os = "wasi", path = "sys/wasi.rs")]
 mod sys;
 
-#[cfg(not(any(windows, unix)))]
+#[cfg(not(any(windows, unix, target_os = "wasi")))]
 compile_error!("Socket2 doesn't support the compile target");
 
 use sys::c_int;
@@ -260,12 +261,12 @@
     /// Type corresponding to `SOCK_STREAM`.
     ///
     /// Used for protocols such as TCP.
-    pub const STREAM: Type = Type(sys::SOCK_STREAM);
+    pub const STREAM: Type = Type(sys::SOCK_STREAM as i32);
 
     /// Type corresponding to `SOCK_DGRAM`.
     ///
     /// Used for protocols such as UDP.
-    pub const DGRAM: Type = Type(sys::SOCK_DGRAM);
+    pub const DGRAM: Type = Type(sys::SOCK_DGRAM as i32);
 
     /// Type corresponding to `SOCK_DCCP`.
     ///
@@ -518,6 +519,7 @@
         target_os = "watchos",
         target_os = "windows",
         target_os = "cygwin",
+        target_os = "wasi",
     ))]
     pub const fn with_interval(self, interval: Duration) -> Self {
         Self {
@@ -542,6 +544,7 @@
             target_os = "visionos",
             target_os = "linux",
             target_os = "macos",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "tvos",
             target_os = "watchos",
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/src/sockaddr.rs b/src/sockaddr.rs
--- a/src/sockaddr.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/sockaddr.rs	2006-07-24 03:21:28.000000000 +0200
@@ -1,6 +1,7 @@
 use std::hash::Hash;
 use std::mem::{self, size_of};
 use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};
+#[cfg(not(target_os = "wasi"))]
 use std::path::Path;
 use std::{fmt, io, ptr};
 
@@ -195,6 +196,7 @@
     /// Constructs a `SockAddr` with the family `AF_UNIX` and the provided path.
     ///
     /// Returns an error if the path is longer than `SUN_LEN`.
+    #[cfg(not(target_os = "wasi"))]
     pub fn unix<P>(path: P) -> io::Result<SockAddr>
     where
         P: AsRef<Path>,
@@ -276,7 +278,7 @@
                 ip,
                 port,
                 addr.sin6_flowinfo,
-                #[cfg(unix)]
+                #[cfg(any(unix, all(target_os = "wasi", target_vendor = "wasmer")))]
                 addr.sin6_scope_id,
                 #[cfg(windows)]
                 unsafe {
@@ -368,7 +370,7 @@
             storage.sin6_port = addr.port().to_be();
             storage.sin6_addr = crate::sys::to_in6_addr(addr.ip());
             storage.sin6_flowinfo = addr.flowinfo();
-            #[cfg(unix)]
+            #[cfg(any(unix, all(target_os = "wasi", target_vendor = "wasmer")))]
             {
                 storage.sin6_scope_id = addr.scope_id();
             }
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/src/socket.rs b/src/socket.rs
--- a/src/socket.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/socket.rs	2006-07-24 03:21:28.000000000 +0200
@@ -16,6 +16,8 @@
 use std::net::{self, Ipv4Addr, Shutdown};
 #[cfg(unix)]
 use std::os::fd::{FromRawFd, IntoRawFd};
+#[cfg(target_os = "wasi")]
+use std::os::wasi::io::{FromRawFd, IntoRawFd};
 #[cfg(windows)]
 use std::os::windows::io::{FromRawSocket, IntoRawSocket};
 use std::time::Duration;
@@ -149,7 +151,7 @@
     /// Creates a pair of sockets which are connected to each other.
     ///
     /// This function corresponds to `socketpair(2)`.
-    #[cfg(all(feature = "all", unix))]
+    #[cfg(all(feature = "all", any(unix, target_os = "wasi")))]
     pub fn pair_raw(
         domain: Domain,
         ty: Type,
@@ -220,7 +222,7 @@
         match res {
             Ok(()) => return Ok(()),
             Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {}
-            #[cfg(unix)]
+            #[cfg(any(unix, target_os = "wasi"))]
             Err(ref e) if e.raw_os_error() == Some(libc::EINPROGRESS) => {}
             Err(e) => return Err(e),
         }
@@ -265,6 +267,7 @@
             target_os = "fuchsia",
             target_os = "illumos",
             target_os = "linux",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "openbsd",
             target_os = "cygwin",
@@ -279,6 +282,7 @@
             target_os = "fuchsia",
             target_os = "illumos",
             target_os = "linux",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "openbsd",
             target_os = "cygwin",
@@ -616,6 +620,7 @@
     /// but suppresses the `WSAEMSGSIZE` error on Windows.
     ///
     /// [`peek_from`]: Socket::peek_from
+    #[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
     pub fn peek_sender(&self) -> io::Result<SockAddr> {
         sys::peek_sender(self.as_raw())
     }
@@ -1165,7 +1170,7 @@
     #[cfg(all(feature = "all", not(any(target_os = "redox", target_os = "espidf"))))]
     pub fn header_included_v4(&self) -> io::Result<bool> {
         unsafe {
-            getsockopt::<c_int>(self.as_raw(), sys::IPPROTO_IP, sys::IP_HDRINCL)
+            getsockopt::<c_int>(self.as_raw(), sys::IPPROTO_IP, libc::IP_HDRINCL)
                 .map(|included| included != 0)
         }
     }
@@ -1191,7 +1196,7 @@
             setsockopt(
                 self.as_raw(),
                 sys::IPPROTO_IP,
-                sys::IP_HDRINCL,
+                libc::IP_HDRINCL,
                 included as c_int,
             )
         }
@@ -1971,7 +1976,7 @@
         target_os = "redox",
         target_os = "solaris",
         target_os = "haiku",
-        target_os = "hurd",
+        target_os = "wasi",
         target_os = "espidf",
         target_os = "vita",
     )))]
@@ -1996,6 +2001,7 @@
         target_os = "redox",
         target_os = "solaris",
         target_os = "haiku",
+        target_os = "wasi",
         target_os = "hurd",
         target_os = "espidf",
         target_os = "vita",
@@ -2124,6 +2130,7 @@
             target_os = "visionos",
             target_os = "linux",
             target_os = "macos",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "tvos",
             target_os = "watchos",
@@ -2154,6 +2161,7 @@
             target_os = "visionos",
             target_os = "linux",
             target_os = "macos",
+            target_os = "wasi",
             target_os = "netbsd",
             target_os = "tvos",
             target_os = "watchos",
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/src/sockref.rs b/src/sockref.rs
--- a/src/sockref.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/sockref.rs	2006-07-24 03:21:28.000000000 +0200
@@ -4,6 +4,8 @@
 use std::ops::Deref;
 #[cfg(unix)]
 use std::os::fd::{AsFd, AsRawFd, FromRawFd};
+#[cfg(target_os = "wasi")]
+use std::os::wasi::io::{AsFd, AsRawFd, FromRawFd};
 #[cfg(windows)]
 use std::os::windows::io::{AsRawSocket, AsSocket, FromRawSocket};
 
@@ -77,7 +79,7 @@
 }
 
 /// On Windows, a corresponding `From<&impl AsSocket>` implementation exists.
-#[cfg(unix)]
+#[cfg(any(unix, target_os = "wasi"))]
 impl<'s, S> From<&'s S> for SockRef<'s>
 where
     S: AsFd,
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/src/sys/unix.rs b/src/sys/unix.rs
--- a/src/sys/unix.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/sys/unix.rs	2006-07-24 03:21:28.000000000 +0200
@@ -2942,4 +2942,4 @@
     let want = [32, 0, 0, 1, 0, 2, 0, 3, 0, 4, 0, 5, 0, 6, 0, 7];
     assert_eq!(raw.s6_addr, want);
     assert_eq!(from_in6_addr(raw), ip);
-}
+}
\ No newline at end of file
