diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/Cargo.toml b/Cargo.toml
--- a/Cargo.toml	2026-07-17 12:56:56.488760674 +0200
+++ b/Cargo.toml	2026-07-17 12:57:37.067296620 +0200
@@ -79,7 +79,7 @@
 [dev-dependencies.x509-parser]
 version = "0.18"
 
-[target.'cfg(all(unix, not(target_os = "macos")))'.dependencies.openssl-probe]
+[target.'cfg(any(all(unix, not(target_os = "macos")), target_os = "wasi"))'.dependencies.openssl-probe]
 version = "0.2"
 
 [target.'cfg(target_os = "macos")'.dependencies.security-framework]
diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' a/src/lib.rs b/src/lib.rs
--- a/src/lib.rs	2026-07-17 12:56:52.626776398 +0200
+++ b/src/lib.rs	2026-07-17 12:57:37.066842160 +0200
@@ -28,9 +28,9 @@
 use pki_types::pem::{self, PemObject};
 use pki_types::CertificateDer;
 
-#[cfg(all(unix, not(target_os = "macos")))]
+#[cfg(any(all(unix, not(target_os = "macos")), target_os = "wasi"))]
 mod unix;
-#[cfg(all(unix, not(target_os = "macos")))]
+#[cfg(any(all(unix, not(target_os = "macos")), target_os = "wasi"))]
 use unix as platform;
 
 #[cfg(windows)]
