--- a/src/rpc/Cargo.toml
+++ b/src/rpc/Cargo.toml
@@ -27,7 +27,6 @@ tracing.workspace = true
 rand.workspace = true
 itertools.workspace = true
 sodiumoxide.workspace = true
-nix.workspace = true
 
 async-trait.workspace = true
 serde.workspace = true
@@ -41,12 +40,14 @@ k8s-openapi = { workspace = true, optional = true }
 schemars = { workspace = true, optional = true }
 reqwest = { workspace = true, optional = true }
 
-pnet_datalink.workspace = true
-
 futures.workspace = true
 tokio.workspace = true
 opentelemetry.workspace = true
 
+[target.'cfg(unix)'.dependencies]
+nix.workspace = true
+pnet_datalink.workspace = true
+
 [features]
 kubernetes-discovery = [ "kube", "k8s-openapi", "schemars" ]
 consul-discovery = [ "reqwest", "thiserror" ]
