--- a/docutils/writers/odf_odt/__init__.py
+++ b/docutils/writers/odf_odt/__init__.py
@@ -1067 +1067 @@
-        except (subprocess.CalledProcessError, FileNotFoundError, ValueError):
+        except (subprocess.CalledProcessError, OSError, ValueError):
--- a/test/test_parsers/test_rst/test_directives/test_include.py
+++ b/test/test_parsers/test_rst/test_directives/test_include.py
@@ -9,0 +10 @@
+import errno
@@ -133,2 +134,2 @@
-    errstr_8bit_path = """\
-InputError: [Errno 2] No such file or directory: '\u043c\u0438\u0440.rst'.\
+    errstr_8bit_path = f"""\
+InputError: [Errno {errno.ENOENT}] No such file or directory: '\u043c\u0438\u0440.rst'.\
@@ -334 +335 @@
-"""\
+f"""\
@@ -342 +343 @@
-                InputError: [Errno 2] No such file or directory: 'nonexistent.rst'.
+                InputError: [Errno {errno.ENOENT}] No such file or directory: 'nonexistent.rst'.
@@ -661 +662 @@
-                InputError: [Errno 2] No such file or directory: '{nonexistent}'.
+                InputError: [Errno {errno.ENOENT}] No such file or directory: '{nonexistent}'.
@@ -943 +944 @@
-            InputError: [Errno 2] No such file or directory: '{nonexistent}'.
+            InputError: [Errno {errno.ENOENT}] No such file or directory: '{nonexistent}'.
--- a/test/test_parsers/test_rst/test_directives/test_raw.py
+++ b/test/test_parsers/test_rst/test_directives/test_raw.py
@@ -10,0 +11 @@
+import errno
@@ -174 +175 @@
-"""\
+f"""\
@@ -179 +180 @@
-            InputError: [Errno 2] No such file or directory: 'non-existent.file'.
+            InputError: [Errno {errno.ENOENT}] No such file or directory: 'non-existent.file'.
--- a/test/test_parsers/test_rst/test_directives/test_tables.py
+++ b/test/test_parsers/test_rst/test_directives/test_tables.py
@@ -10,0 +11 @@
+import errno
@@ -899 +900 @@
-"""\
+f"""\
@@ -904 +905 @@
-            [Errno 2] No such file or directory: 'bogus.csv'.
+            [Errno {errno.ENOENT}] No such file or directory: 'bogus.csv'.
--- a/test/test_writers/test_html5_polyglot.py
+++ b/test/test_writers/test_html5_polyglot.py
@@ -12,0 +13 @@
+import errno
@@ -606 +607 @@
-  [Errno 2] No such file or directory: 'dummy.png'</p>
+  [Errno {errno.ENOENT}] No such file or directory: 'dummy.png'</p>
--- a/test/test_writers/test_latex2e_parts.py
+++ b/test/test_writers/test_latex2e_parts.py
@@ -20,0 +21 @@
+import errno
@@ -820 +821 @@
-%  [Errno 2] No such file or directory: {spampath}
+%  [Errno {errno.ENOENT}] No such file or directory: {spampath}
