diff --git a/configure.ac b/configure.ac
index a3fdd630..e37aab27 100644
--- a/configure.ac
+++ b/configure.ac
@@ -548,6 +548,7 @@ AC_CHECK_FUNCS(m4_normalize([
   funopen
   gai_strerror
   getcwd
+  getdtablesize
   getgrnam_r
   gethostname
   getloadavg
diff --git a/ext/standard/php_fopen_wrapper.c b/ext/standard/php_fopen_wrapper.c
index ea33ba49..6c04c82c 100644
--- a/ext/standard/php_fopen_wrapper.c
+++ b/ext/standard/php_fopen_wrapper.c
@@ -317,7 +317,7 @@ static php_stream * php_stream_url_wrap_php(php_stream_wrapper *wrapper, const c
 			return NULL;
 		}
 
-#ifdef HAVE_UNISTD_H
+#ifdef HAVE_GETDTABLESIZE
 		dtablesize = getdtablesize();
 #else
 		dtablesize = INT_MAX;
