diff --git a/ext/random/csprng.c b/ext/random/csprng.c
--- a/ext/random/csprng.c
+++ b/ext/random/csprng.c
@@ -45,7 +45,8 @@
 #if HAVE_SYS_PARAM_H
 # include <sys/param.h>
 # if (__FreeBSD__ && __FreeBSD_version > 1200000) || (__DragonFly__ && __DragonFly_version >= 500700) || \
-     (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000)
+     (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000) || \
+     (defined(__wasi__) && defined(HAVE_GETRANDOM))
 #  include <sys/random.h>
 # endif
 #endif
@@ -98,7 +99,8 @@ PHPAPI zend_result php_random_bytes(void *bytes, size_t size, bool should_throw)
 #else
 	size_t read_bytes = 0;
 # if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || \
-  (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000)
+  (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__NetBSD__) && __NetBSD_Version__ >= 1000000000) || \
+  (defined(__wasi__) && defined(HAVE_GETRANDOM))
 	/* Linux getrandom(2) syscall or FreeBSD/DragonFlyBSD/NetBSD getrandom(2) function
 	 * Being a syscall, implemented in the kernel, getrandom offers higher quality output
 	 * compared to the arc4random api albeit a fallback to /dev/urandom is considered.
