diff --git a/ext/random/random.c b/ext/random/random.c
--- a/ext/random/random.c
+++ b/ext/random/random.c
@@ -47,7 +47,8 @@
 
 #if HAVE_SYS_PARAM_H
 # include <sys/param.h>
-# if (__FreeBSD__ && __FreeBSD_version > 1200000) || (__DragonFly__ && __DragonFly_version >= 500700) || (defined(__sun) && defined(HAVE_GETRANDOM))
+# if (__FreeBSD__ && __FreeBSD_version > 1200000) || (__DragonFly__ && __DragonFly_version >= 500700) || \
+     (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__wasi__) && defined(HAVE_GETRANDOM))
 #  include <sys/random.h>
 # endif
 #endif
@@ -511,7 +512,8 @@ PHPAPI zend_result php_random_bytes(void *bytes, size_t size, bool should_throw)
 #else
 	size_t read_bytes = 0;
 	ssize_t n;
-# if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || (defined(__sun) && defined(HAVE_GETRANDOM))
+# if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || (defined(__DragonFly__) && __DragonFly_version >= 500700) || \
+  (defined(__sun) && defined(HAVE_GETRANDOM)) || (defined(__wasi__) && defined(HAVE_GETRANDOM))
 	/* Linux getrandom(2) syscall or FreeBSD/DragonFlyBSD getrandom(2) function*/
 	/* Keep reading until we get enough entropy */
 	while (read_bytes < size) {
