diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -610,6 +610,7 @@ getprotobynumber \
 getservbyname \
 getservbyport \
 getrusage \
+getrandom \
 gettimeofday \
 gmtime_r \
 getpwnam_r \
diff --git a/ext/standard/random.c b/ext/standard/random.c
--- a/ext/standard/random.c
+++ b/ext/standard/random.c
@@ -37,6 +37,9 @@
 #  include <sys/random.h>
 # endif
 #endif
+#if defined(__wasi__) && defined(HAVE_GETRANDOM)
+# include <sys/random.h>
+#endif
 
 #if __has_feature(memory_sanitizer)
 # include <sanitizer/msan_interface.h>
@@ -103,7 +106,8 @@ PHPAPI int php_random_bytes(void *bytes, size_t size, zend_bool should_throw)
 #else
 	size_t read_bytes = 0;
 	ssize_t n;
-#if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000)
+#if (defined(__linux__) && defined(SYS_getrandom)) || (defined(__FreeBSD__) && __FreeBSD_version >= 1200000) || \
+    (defined(__wasi__) && defined(HAVE_GETRANDOM))
 	/* Linux getrandom(2) syscall or FreeBSD getrandom(2) function*/
 	/* Keep reading until we get enough entropy */
 	while (read_bytes < size) {
