--- a/src/credential_process.rs
+++ b/src/credential_process.rs
@@ -85,7 +85,7 @@ impl CredentialProcessProvider {
         // Security: command arguments must be redacted at debug level
         tracing::debug!(command = %self.command, "loading credentials from external process");
 
-        let command = if cfg!(windows) {
+        let mut command = if cfg!(windows) {
             let mut command = Command::new("cmd.exe");
             command.args(["/C", self.command.unredacted()]);
             command
@@ -95,10 +95,24 @@ impl CredentialProcessProvider {
             command
         };
+        #[cfg(not(all(target_os = "wasi", target_vendor = "wasmer")))]
         let output = tokio::process::Command::from(command)
             .output()
             .await
             .map_err(|e| {
                 CredentialsError::provider_error(format!(
                     "Error retrieving credentials from external process: {e}",
                 ))
             })?;
+        #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+        let output = tokio::task::spawn_blocking(move || command.output())
+            .await
+            .map_err(|e| {
+                CredentialsError::provider_error(format!(
+                    "Error waiting for external credentials process: {e}",
+                ))
+            })?
+            .map_err(|e| {
+                CredentialsError::provider_error(format!(
+                    "Error retrieving credentials from external process: {e}",
+                ))
+            })?;
