diff --git a/Cargo.toml b/Cargo.toml
index 517f4fe..79bae6f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -75,6 +75,16 @@ features = [
     "net",
 ]
 
+[features]
+default = ["log"]
+net = []
+os-ext = [
+    "os-poll",
+    "windows-sys/Win32_System_Pipes",
+    "windows-sys/Win32_Security",
+]
+os-poll = []
+
 [lib]
 name = "mio"
 path = "src/lib.rs"
@@ -114,27 +124,23 @@ default-features = false
 [dev-dependencies.rand]
 version = "0.8"
 
-[features]
-default = ["log"]
-net = []
-os-ext = [
-    "os-poll",
-    "windows-sys/Win32_System_Pipes",
-    "windows-sys/Win32_Security",
-]
-os-poll = []
+[target.'cfg(all(target_os = "wasi", target_vendor = "unknown"))'.dependencies.wasi]
+version = "^0.11"
+
+[target.'cfg(all(target_os = "wasi", target_vendor = "wasmer"))'.dependencies.wasix]
+version = "0.13"
 
 [target.'cfg(target_os = "hermit")'.dependencies.libc]
-version = "0.2.159"
+version = "0.2.169"
 
 [target.'cfg(target_os = "wasi")'.dependencies.libc]
-version = "0.2.159"
+version = "0.2.169"
 
 [target.'cfg(target_os = "wasi")'.dependencies.wasi]
 version = "0.11.0"
 
 [target."cfg(unix)".dependencies.libc]
-version = "0.2.159"
+version = "0.2.169"
 
 [target."cfg(windows)".dependencies.windows-sys]
 version = "0.52"
diff --git a/src/lib.rs b/src/lib.rs
index 8004ee2..d5cea68 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -40,6 +40,10 @@
 //!
 //! The available features are described in the [`features`] module.
 
+// WASIX requires os-poll when using the feature net
+#[cfg(all(target_os = "wasi", target_vendor = "wasmer", feature = "net", not(feature = "os-poll")))]
+compile_error!("WASIX requires os-poll when using the feature net");
+
 // macros used internally
 #[macro_use]
 mod macros;
@@ -48,8 +52,10 @@ mod interest;
 mod poll;
 mod sys;
 mod token;
-#[cfg(not(target_os = "wasi"))]
-mod waker;
+
+cfg_os_poll! {
+    mod waker;
+}
 
 pub mod event;
 
@@ -66,14 +72,25 @@ pub use event::Events;
 pub use interest::Interest;
 pub use poll::{Poll, Registry};
 pub use token::Token;
-#[cfg(not(target_os = "wasi"))]
-pub use waker::Waker;
 
-#[cfg(all(unix, feature = "os-ext"))]
-#[cfg_attr(docsrs, doc(cfg(all(unix, feature = "os-ext"))))]
+mod wasi {
+    #[allow(unused)]
+    #[cfg(all(target_os = "wasi", target_vendor = "unknown"))]
+    pub use ::wasi::*;
+    #[allow(unused)]
+    #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+    pub use ::wasix::*;
+}
+
+cfg_os_poll! {
+    pub use waker::Waker;
+}
+
+#[cfg(all(any(unix, target_vendor = "wasmer"), feature = "os-ext"))]
+#[cfg_attr(docsrs, doc(cfg(all(any(unix, target_vendor = "wasmer"), feature = "os-ext"))))]
 pub mod unix {
     //! Unix only extensions.
-
+    #[cfg(unix)]
     pub mod pipe {
         //! Unix pipe.
         //!
@@ -82,6 +99,7 @@ pub mod unix {
         pub use crate::sys::pipe::{new, Receiver, Sender};
     }
 
+    #[allow(unused)]
     pub use crate::sys::SourceFd;
 }
 
diff --git a/src/net/mod.rs b/src/net/mod.rs
index 15c405c..b1c2ddc 100644
--- a/src/net/mod.rs
+++ b/src/net/mod.rs
@@ -28,9 +28,9 @@
 mod tcp;
 pub use self::tcp::{TcpListener, TcpStream};
 
-#[cfg(not(target_os = "wasi"))]
+#[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
 mod udp;
-#[cfg(not(target_os = "wasi"))]
+#[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
 pub use self::udp::UdpSocket;
 
 #[cfg(unix)]
diff --git a/src/net/tcp/listener.rs b/src/net/tcp/listener.rs
index 2e27a76..8b99aac 100644
--- a/src/net/tcp/listener.rs
+++ b/src/net/tcp/listener.rs
@@ -13,9 +13,13 @@ use std::{fmt, io};
 
 use crate::io_source::IoSource;
 use crate::net::TcpStream;
-#[cfg(any(unix, target_os = "hermit"))]
+#[cfg(any(
+    unix,
+    target_os = "hermit",
+    all(target_os = "wasi", target_vendor = "wasmer")
+))]
 use crate::sys::tcp::set_reuseaddr;
-#[cfg(not(target_os = "wasi"))]
+#[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
 use crate::sys::tcp::{bind, listen, new_for_addr};
 use crate::{event, sys, Interest, Registry, Token};
 
@@ -59,11 +63,13 @@ impl TcpListener {
     /// 2. Set the `SO_REUSEADDR` option on the socket on Unix.
     /// 3. Bind the socket to the specified address.
     /// 4. Calls `listen` on the socket to prepare it to receive new connections.
-    #[cfg(not(target_os = "wasi"))]
+    #[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
     pub fn bind(addr: SocketAddr) -> io::Result<TcpListener> {
         let socket = new_for_addr(addr)?;
         #[cfg(any(unix, target_os = "hermit"))]
         let listener = unsafe { TcpListener::from_raw_fd(socket) };
+        #[cfg(target_os = "wasi")]
+        let listener = unsafe { TcpListener::from_raw_fd(socket) };
         #[cfg(windows)]
         let listener = unsafe { TcpListener::from_raw_socket(socket as _) };
 
diff --git a/src/net/tcp/stream.rs b/src/net/tcp/stream.rs
index f3db74f..04bbf51 100644
--- a/src/net/tcp/stream.rs
+++ b/src/net/tcp/stream.rs
@@ -13,7 +13,7 @@ use std::os::windows::io::{
 };
 
 use crate::io_source::IoSource;
-#[cfg(not(target_os = "wasi"))]
+#[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
 use crate::sys::tcp::{connect, new_for_addr};
 use crate::{event, Interest, Registry, Token};
 
@@ -82,16 +82,19 @@ impl TcpStream {
     ///     went wrong.
     ///  6. Now the stream can be used.
     ///
+    /// [read interest]: Interest::READABLE
     /// This may return a `WouldBlock` in which case the socket connection
     /// cannot be completed immediately, it usually means there are insufficient
     /// entries in the routing cache.
     ///
     /// [write interest]: Interest::WRITABLE
-    #[cfg(not(target_os = "wasi"))]
+    #[cfg(not(all(target_os = "wasi", target_vendor = "unknown")))]
     pub fn connect(addr: SocketAddr) -> io::Result<TcpStream> {
         let socket = new_for_addr(addr)?;
         #[cfg(any(unix, target_os = "hermit"))]
         let stream = unsafe { TcpStream::from_raw_fd(socket) };
+        #[cfg(target_os = "wasi")]
+        let stream = unsafe { TcpStream::from_raw_fd(socket) };
         #[cfg(windows)]
         let stream = unsafe { TcpStream::from_raw_socket(socket as _) };
         connect(&stream.inner, addr)?;
diff --git a/src/poll.rs b/src/poll.rs
index a175cdd..3a52d9c 100644
--- a/src/poll.rs
+++ b/src/poll.rs
@@ -15,7 +15,7 @@
 ))]
 use std::os::fd::{AsRawFd, RawFd};
 #[cfg(all(debug_assertions, not(target_os = "wasi")))]
-use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::atomic::AtomicBool;
 #[cfg(all(debug_assertions, not(target_os = "wasi")))]
 use std::sync::Arc;
 use std::time::Duration;
@@ -717,17 +717,22 @@ impl Registry {
 
     /// Internal check to ensure only a single `Waker` is active per [`Poll`]
     /// instance.
-    #[cfg(all(debug_assertions, not(target_os = "wasi")))]
-    pub(crate) fn register_waker(&self) {
+    #[allow(dead_code)]
+    #[cfg(all(debug_assertions))]
+    pub fn register_waker(&self) {
+        #[cfg(all(debug_assertions, not(target_os = "wasi")))]
         assert!(
-            !self.has_waker.swap(true, Ordering::AcqRel),
+            !self
+                .has_waker
+                .swap(true, std::sync::atomic::Ordering::AcqRel),
             "Only a single `Waker` can be active per `Poll` instance"
         );
     }
 
     /// Get access to the `sys::Selector`.
-    #[cfg(any(not(target_os = "wasi"), feature = "net"))]
-    pub(crate) fn selector(&self) -> &sys::Selector {
+    #[allow(dead_code)]
+    #[cfg(any(not(target_os = "wasi"), feature = "net", feature = "os-poll"))]
+    pub fn selector(&self) -> &sys::Selector {
         &self.selector
     }
 }
diff --git a/src/sys/mod.rs b/src/sys/mod.rs
index 5f1ee9c..af0a475 100644
--- a/src/sys/mod.rs
+++ b/src/sys/mod.rs
@@ -15,6 +15,7 @@
 //! * `Waker`: see [`crate::Waker`].
 
 cfg_os_poll! {
+    #[allow(unused_macros)]
     macro_rules! debug_detail {
         (
             $type: ident ($event_type: ty), $test: path,
@@ -68,12 +69,28 @@ cfg_os_poll! {
 cfg_os_poll! {
     mod wasi;
     pub(crate) use self::wasi::*;
+
+    #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+    pub use self::wasi::SourceFd;
+
+    #[allow(unused_imports)]
+    #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+    pub use std::net::SocketAddr;
 }
 
 cfg_not_os_poll! {
     mod shell;
     pub(crate) use self::shell::*;
 
+    #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+    cfg_any_os_ext! {
+        mod wasi;
+        pub use self::wasi::SourceFd;
+    }
+    #[allow(unused_imports)]
+    #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+    pub use std::net::SocketAddr;
+
     #[cfg(unix)]
     cfg_any_os_ext! {
         mod unix;
diff --git a/src/sys/shell/mod.rs b/src/sys/shell/mod.rs
index aa1c622..5277bf8 100644
--- a/src/sys/shell/mod.rs
+++ b/src/sys/shell/mod.rs
@@ -9,6 +9,7 @@ pub(crate) use self::selector::{event, Event, Events, Selector};
 
 #[cfg(not(target_os = "wasi"))]
 mod waker;
+#[allow(unused_imports)]
 #[cfg(not(target_os = "wasi"))]
 pub(crate) use self::waker::Waker;
 
diff --git a/src/sys/shell/selector.rs b/src/sys/shell/selector.rs
index 83456ef..89f8b63 100644
--- a/src/sys/shell/selector.rs
+++ b/src/sys/shell/selector.rs
@@ -18,6 +18,12 @@ impl Selector {
     pub fn select(&self, _: &mut Events, _: Option<Duration>) -> io::Result<()> {
         os_required!();
     }
+
+    #[allow(dead_code)]
+    #[cfg(all(debug_assertions))]
+    pub fn register_waker(&self) -> bool {
+        os_required!();
+    }
 }
 
 #[cfg(unix)]
diff --git a/src/sys/shell/waker.rs b/src/sys/shell/waker.rs
index bbdd7c3..c27717e 100644
--- a/src/sys/shell/waker.rs
+++ b/src/sys/shell/waker.rs
@@ -6,10 +6,12 @@ use std::io;
 pub struct Waker {}
 
 impl Waker {
+    #[allow(dead_code)]
     pub fn new(_: &Selector, _: Token) -> io::Result<Waker> {
         os_required!();
     }
 
+    #[allow(dead_code)]
     pub fn wake(&self) -> io::Result<()> {
         os_required!();
     }
diff --git a/src/sys/unix/net.rs b/src/sys/unix/net.rs
index 82d4bb5..cb0e2cd 100644
--- a/src/sys/unix/net.rs
+++ b/src/sys/unix/net.rs
@@ -20,6 +20,7 @@ pub(crate) fn new_socket(domain: libc::c_int, socket_type: libc::c_int) -> io::R
         target_os = "hurd",
         target_os = "illumos",
         target_os = "linux",
+        target_os = "wasi",
         target_os = "netbsd",
         target_os = "openbsd",
         target_os = "solaris",
diff --git a/src/sys/unix/pipe.rs b/src/sys/unix/pipe.rs
index 8e86753..a8b1d58 100644
--- a/src/sys/unix/pipe.rs
+++ b/src/sys/unix/pipe.rs
@@ -15,6 +15,7 @@ pub(crate) fn new_raw() -> io::Result<[RawFd; 2]> {
         target_os = "fuchsia",
         target_os = "hurd",
         target_os = "linux",
+        target_os = "wasi",
         target_os = "netbsd",
         target_os = "openbsd",
         target_os = "illumos",
diff --git a/src/sys/unix/tcp.rs b/src/sys/unix/tcp.rs
index f8dfa58..61655d9 100644
--- a/src/sys/unix/tcp.rs
+++ b/src/sys/unix/tcp.rs
@@ -72,6 +72,7 @@ pub(crate) fn accept(listener: &net::TcpListener) -> io::Result<(net::TcpStream,
         target_os = "hurd",
         target_os = "illumos",
         target_os = "linux",
+        target_os = "wasi",
         target_os = "netbsd",
         target_os = "openbsd",
         target_os = "solaris",
