diff -ruN a/src/sys/wasi/epoll.rs b/src/sys/wasi/epoll.rs
--- a/src/sys/wasi/epoll.rs	2026-08-20 23:24:29.014223138 +0200
+++ b/src/sys/wasi/epoll.rs	2026-08-21 11:41:57.190346407 +0200
@@ -1,6 +1,5 @@
 use crate::{Interest, Token};
 
-use log::error;
 use std::os::fd::AsRawFd;
 use std::os::fd::RawFd;
 #[cfg(debug_assertions)]
@@ -121,7 +120,8 @@
 impl Drop for Selector {
     fn drop(&mut self) {
         if let Err(err) = unsafe { wasi::fd_close(self.ep).map(|_| ()) } {
-            error!("error closing epoll: {}", err);
+            #[cfg(feature = "log")]
+            log::error!("error closing epoll: {}", err);
         }
     }
 }
diff -ruN a/src/sys/wasi/sourcefd.rs b/src/sys/wasi/sourcefd.rs
--- a/src/sys/wasi/sourcefd.rs	2026-08-20 23:24:29.014223138 +0200
+++ b/src/sys/wasi/sourcefd.rs	2026-08-21 11:41:57.191346388 +0200
@@ -3,6 +3,8 @@
 use std::io;
 use std::os::wasi::io::RawFd;
 
+/// Adapter for [`RawFd`] providing an [`event::Source`] implementation.
+///
 #[derive(Debug)]
 pub struct SourceFd<'a>(pub &'a RawFd);
 
