diff -ruN a/src/sys/wasi/mod.rs b/src/sys/wasi/mod.rs
--- a/src/sys/wasi/mod.rs	2026-08-20 23:24:29.014223138 +0200
+++ b/src/sys/wasi/mod.rs	2026-08-21 11:41:57.016349848 +0200
@@ -12,7 +12,6 @@
 //! work. However this could be implemented by use of an `Arc`.
 //!
 //! In summary, this only (barely) works using a single thread.
-
 use std::io;
 
 #[cfg(all(feature = "net", target_vendor = "unknown"))]
diff -ruN a/src/sys/wasi/poll.rs b/src/sys/wasi/poll.rs
--- a/src/sys/wasi/poll.rs	2026-08-20 23:24:29.014223138 +0200
+++ b/src/sys/wasi/poll.rs	2026-08-21 11:41:57.024349690 +0200
@@ -264,7 +264,7 @@
                     // Return the events
                     check_errors(&events)
                 }
-                Err(err) => Err(io_err(err)),
+                Err(err) => Err(io_err(err.into())),
             };
         }
     }
@@ -429,7 +429,7 @@
 fn check_errors(events: &[Event]) -> io::Result<()> {
     for event in events {
         if event.error.raw() != wasi::ERRNO_SUCCESS.raw() {
-            return Err(io_err(event.error));
+            return Err(io_err(event.error.into()));
         }
     }
     Ok(())
