diff --git a/Zend/zend_vm_gen.php b/Zend/zend_vm_gen.php
index 38c20b24da2..9df053312d4 100755
--- a/Zend/zend_vm_gen.php
+++ b/Zend/zend_vm_gen.php
@@ -2498,7 +2498,7 @@ function gen_vm_opcodes_header(
         $str .= "# define ZEND_VM_KIND\t\tZEND_VM_KIND_HYBRID\n";
     }
     if ($GLOBALS["vm_kind_name"][ZEND_VM_GEN_KIND] === "ZEND_VM_KIND_HYBRID" || $GLOBALS["vm_kind_name"][ZEND_VM_GEN_KIND] === "ZEND_VM_KIND_CALL") {
-        $str .= "#elif defined(HAVE_MUSTTAIL) && defined(HAVE_PRESERVE_NONE) && (defined(__x86_64__) || defined(__aarch64__))\n";
+        $str .= "#elif defined(HAVE_MUSTTAIL) && ((defined(HAVE_PRESERVE_NONE) && (defined(__x86_64__) || defined(__aarch64__))) || defined(__wasm32__))\n";
         $str .= "# define ZEND_VM_KIND\t\tZEND_VM_KIND_TAILCALL\n";
         $str .= "#else\n";
         $str .= "# define ZEND_VM_KIND\t\tZEND_VM_KIND_CALL\n";
@@ -2515,8 +2515,12 @@ function gen_vm_opcodes_header(
     $str .= "#endif\n";
     $str .= "\n";
     $str .= "#if ZEND_VM_KIND == ZEND_VM_KIND_TAILCALL\n";
-    $str .= "# define ZEND_OPCODE_HANDLER_CCONV    ZEND_PRESERVE_NONE\n";
+    $str .= "# ifdef HAVE_PRESERVE_NONE\n";
+    $str .= "#  define ZEND_OPCODE_HANDLER_CCONV    ZEND_PRESERVE_NONE\n";
+    $str .= "# else\n";
+    $str .= "#  define ZEND_OPCODE_HANDLER_CCONV    ZEND_FASTCALL\n";
+    $str .= "# endif\n";
     $str .= "# define ZEND_OPCODE_HANDLER_CCONV_EX ZEND_FASTCALL\n";
     $str .= "#elif ZEND_VM_KIND == ZEND_VM_KIND_CALL\n";
     $str .= "# define ZEND_OPCODE_HANDLER_CCONV    ZEND_FASTCALL\n";
