diff --git a/Cargo.toml b/Cargo.toml
index 48ada6b..11644c9 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -128,6 +128,9 @@ features = [
 ]
 default-features = false

+[target.'cfg(all(target_os = "wasi", target_vendor = "wasmer"))'.dependencies.wasix]
+version = "0.13"
+
 [target.'cfg(target_os = "hermit")'.dependencies.hermit-abi]
 version = "0.5.0"

diff --git a/src/lib.rs b/src/lib.rs
index 768ba32..3b64064 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -109,8 +109,11 @@ cfg_if! {
         target_os = "horizon",
         unix,
     ))] {
         mod poll;
         use poll as sys;
+    } else if #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))] {
+        mod wasix;
+        use wasix as sys;
     } else if #[cfg(target_os = "windows")] {
         mod iocp;
         use iocp as sys;
@@ -985,5 +988,6 @@ impl fmt::Debug for Events {
         target_os = "openbsd",
         target_os = "dragonfly",
+        all(target_os = "wasi", target_vendor = "wasmer"),
     ),
     not(polling_test_poll_backend),
 ))]
@@ -1003,8 +1007,11 @@ mod raw_fd_impl {
     )))
 )]
 mod raw_fd_impl {
     use crate::Poller;
+    #[cfg(unix)]
     use std::os::unix::io::{AsFd, AsRawFd, BorrowedFd, RawFd};
+    #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+    use std::os::wasi::io::{AsFd, AsRawFd, BorrowedFd, RawFd};

     impl AsRawFd for Poller {
         fn as_raw_fd(&self) -> RawFd {
@@ -1044,11 +1051,13 @@ impl fmt::Debug for Poller {
 }

 cfg_if! {
-    if #[cfg(any(unix, target_os = "hermit"))] {
+    if #[cfg(any(unix, target_os = "hermit", all(target_os = "wasi", target_vendor = "wasmer")))] {
         #[cfg(unix)]
         use std::os::unix::io::{AsRawFd, RawFd, AsFd, BorrowedFd};
         #[cfg(target_os = "hermit")]
         use std::os::hermit::io::{AsRawFd, RawFd, AsFd, BorrowedFd};
+        #[cfg(all(target_os = "wasi", target_vendor = "wasmer"))]
+        use std::os::wasi::io::{AsRawFd, RawFd, AsFd, BorrowedFd};

         /// A resource with a raw file descriptor.
         pub trait AsRawSource {
