diff --git a/include/clang/Support/Compiler.h b/include/clang/Support/Compiler.h
index 38e7f127ae33..92a00a3d5af7 100644
--- a/include/clang/Support/Compiler.h
+++ b/include/clang/Support/Compiler.h
@@ -52,7 +52,8 @@
 #define CLANG_ABI LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
 #define CLANG_TEMPLATE_ABI LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
 #define CLANG_EXPORT_TEMPLATE
-#elif defined(__MACH__) || defined(__WASM__) || defined(__EMSCRIPTEN__)
+#elif defined(__MACH__) || defined(__wasm__) || defined(__WASM__) ||          \
+    defined(__EMSCRIPTEN__)
 #define CLANG_ABI LLVM_ATTRIBUTE_VISIBILITY_DEFAULT
 #define CLANG_TEMPLATE_ABI
 #define CLANG_EXPORT_TEMPLATE
