diff -ruN '--exclude=.cargo-checksum.json' '--exclude=Cargo.lock' '--exclude=target' '--exclude=*.orig' '--exclude=*.rej' a/src/system_conf/unix.rs b/src/system_conf/unix.rs
--- a/src/system_conf/unix.rs	2006-07-24 03:21:28.000000000 +0200
+++ b/src/system_conf/unix.rs	2006-07-24 03:21:28.000000000 +0200
@@ -51,7 +51,10 @@
 fn into_resolver_config(
     parsed_config: resolv_conf::Config,
 ) -> io::Result<(ResolverConfig, ResolverOpts)> {
-    let domain = if let Some(domain) = parsed_config.get_system_domain() {
+    // let domain = if let Some(domain) = parsed_config.get_system_domain() {
+    // HACK: Fix the domain name to WASIX's domain name
+    let system_domain = Some("wasix.local".to_owned());
+    let domain = if let Some(domain) = system_domain {
         // The system domain name maybe appear to be valid to the resolv_conf
         // crate but actually be invalid. For example, if the hostname is "matt.schulte's computer"
         // In order to prevent a hostname which macOS or Windows would consider
