diff --git a/Zend/zend_alloc.c b/Zend/zend_alloc.c
--- a/Zend/zend_alloc.c
+++ b/Zend/zend_alloc.c
@@ -698,7 +698,7 @@ static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
 	if (ptr == NULL) {
 		return NULL;
 	} else if (ZEND_MM_ALIGNED_OFFSET(ptr, alignment) == 0) {
-#ifdef MADV_HUGEPAGE
+#if defined(MADV_HUGEPAGE) && !defined(__wasi__)
 		if (zend_mm_use_huge_pages) {
 			madvise(ptr, size, MADV_HUGEPAGE);
 		}
@@ -739,7 +739,7 @@ static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
 		if (alignment > REAL_PAGE_SIZE) {
 			zend_mm_munmap((char*)ptr + size, alignment - REAL_PAGE_SIZE);
 		}
-# ifdef MADV_HUGEPAGE
+# if defined(MADV_HUGEPAGE) && !defined(__wasi__)
 		if (zend_mm_use_huge_pages) {
 			madvise(ptr, size, MADV_HUGEPAGE);
 		}
