diff --git a/src/compiler/flags.rs b/src/compiler/flags.rs
index 2d164bc..66b3564 100644
--- a/src/compiler/flags.rs
+++ b/src/compiler/flags.rs
@@ -245,6 +245,7 @@ fn is_linker_input(path: &str) -> bool {
     if matches!(
         Path::new(path).extension().and_then(|ext| ext.to_str()),
-        Some("o") | Some("obj") | Some("a") | Some("so") | Some("dll") | Some("dylib")
+        Some("o") | Some("obj") | Some("a") | Some("rlib") | Some("so") | Some("dll")
+            | Some("dylib")
     ) {
         return true;
     }
@@ -531,6 +532,7 @@ mod tests {
             "foo.o",
             "foo.obj",
             "libfoo.a",
+            "libfoo.rlib",
             "libfoo.so",
             "foo.dll",
             "libfoo.dylib",
